{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NOTEBOOK_HEADER-->\n",
    "*This notebook contains material from [PyRosetta](https://RosettaCommons.github.io/PyRosetta);\n",
    "content is available [on Github](https://github.com/RosettaCommons/PyRosetta.notebooks.git).*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [Side Chain Conformations and Dunbrack Energies](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.01-Side-Chain-Conformations-and-Dunbrack-Energies.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [Protein Design with a Resfile and FastRelax](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.03-Design-with-a-resfile-and-relax.ipynb) ><p><a href=\"https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.02-Packing-design-and-regional-relax.ipynb\"><img align=\"left\" src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open in Colab\" title=\"Open in Google Colaboratory\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RosettaAntibodyDesign\n",
    "## Notes\n",
    "This tutorial will walk you through how to use `RosettaAntibodyDesign` in PyRosetta.  You should also go through the parellel distribution workshop as you will most likely need to create many decoys for some of these design tasks. Note that we are using the XML interface to the code here for simplicity (and because I had a C++ workshop I am converting - truth be told).  The code-level interface is as robust as the XML - but will require more knowledge use.  You are welcome to play around with it - all functions have descriptions and all options are possible to change through code.\n",
    "\n",
    "Grab a coffee, take a breath, and lets learn how to design some antibodies!\n",
    "\n",
    "## Citation\n",
    "\n",
    "\n",
    "[Rosetta Antibody Design (RAbD): A General Framework for Computational Antibody Design, PLOS Computational Biology, 4/27/2018](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006112)\n",
    "\n",
    "Jared Adolf-Bryfogle, Oleks Kalyuzhniy, Michael Kubitz, Brian D. Weitzner, Xiaozhen Hu, Yumiko Adachi, William R. Schief, Roland L. Dunbrack Jr."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "__RosettaAntibodyDesign (RAbD)__ is a generalized framework for the design of antibodies, in which a user can easily tailor the run to their project needs.  __The algorithm is meant to sample the diverse sequence, structure, and binding space of an antibody-antigen complex.__  An app is available, and all components can be used within RosettaScripts for easy scripting of antibody design and incorporation into other Rosetta protocols.\n",
    "\n",
    "The framework is based on rigorous bioinformatic analysis and rooted very much on our [recent clustering](https://www.ncbi.nlm.nih.gov/pubmed/21035459) of antibody CDR regions.  It uses the __North/Dunbrack CDR definition__ as outlined in the North/Dunbrack clustering paper. A new clustering paper will be out in the next year, and this new analysis will be incorporated into RAbD. \n",
    "\n",
    "The supplemental methods section of the published paper has all details of the RosettaAntibodyDesign method.  This manual serves to get you started running RAbD in typical use fashions. \n",
    "\n",
    "\n",
    "# Algorithm\n",
    "  \n",
    "Broadly, the RAbD protocol consists of alternating outer and inner Monte Carlo cycles. Each outer cycle consists of randomly choosing a CDR (L1, L2, etc.) from those CDRs set to design, randomly choosing a cluster and then a structure from that cluster from the database according to the input instructions, and optionally grafting that CDR's structure onto the antibody framework in place of the existing CDR (__GraftDesign__). The program then performs N rounds of the inner cycle, consisting of sequence design (__SeqDesign__) using cluster-based sequence profiles and structural constraints, energy minimization, and optional docking. Each inner cycle structurally optimizes the backbone and repacks side chains of the CDR chosen in the outer cycle as well as optional neighbors in order to optimize interactions of the CDR with the antigen and other CDRs. \n",
    "\n",
    "__Backbone dihedral angle (CircularHarmonic) constraints__ derived from the cluster data are applied to each CDR to limit deleterious structural perturbations. Amino acid changes are typically sampled from __profiles derived for each CDR cluster in PyIgClassify__. Conservative amino acid substitutions (according to the BLOSUM62 substitution matrix) may be performed when too few sequences are available to produce a profile (e.g., for H3). After each inner cycle is completed, the new sequence and structure are accepted according to the Metropolis Monte Carlo criterion. After N rounds within the inner cycle, the program returns to the outer cycle, at which point the energy of the resulting design is compared to the previous design in the outer cycle. The new design is accepted or rejected according to the Monte Carlo criterion.\n",
    "\n",
    "If optimizing the antibody-antigen orientation during the design (dock), SiteConstraints are automatically used to keep the CDRs (paratope) facing the antigen surface.  These are termed __ParatopeSiteConstraints__.   Optionally, one can enable constraints that keep the paratope of the antibody around a target epitope (antigen binding site).  These are called __ParatopeEpitopeSiteConstraints__ as the constraints are between the paratope and the epitope. The epitope is automatically determined as the interface residues around the paratope on input into the program, however, any residue(s) can be set as the epitope to limit unwanted movement and sampling of the antibody.  See the examples and options below. \n",
    "\n",
    "More detail on the algorithm can be found in the published paper. \n",
    "\n",
    "\n",
    "# General Setup and Inputs\n",
    "\n",
    "1.  Antibody Design Database\n",
    "\n",
    "\tThis app requires the Rosetta Antibody Design Database.  A database of antibodies from the original North Clustering paper is included in Rosetta and is used as the default .  An updated database (which is currently updated bi-yearly) can be downloaded here: <http://dunbrack2.fccc.edu/PyIgClassify/>.  \n",
    "\n",
    "\tFor C++, It should be placed in `Rosetta/main/database/sampling/antibodies/`. For PyRosetta, use the cmd-line option `antibody_database` and set it to the full path of the downloaded database within the `init()` function as you have done in the past. It is recommended to use this up-to-date database for production runs. For this tutorial, we will use the database within Rosetta. \n",
    "\n",
    "\n",
    "2.  Starting Structure\n",
    "\n",
    "\tThe protocol begins with the three-dimensional structure of an antibody-antigen complex. Designs should start with an antibody bound to a target antigen (however optimizing just the antibody without the complex is also possible).  Camelid antibodies are fully supported.  This structure may be an experimental structure of an existing antibody in complex with its antigen, a predicted structure of an existing antibody docked computationally to its antigen, or even the best scoring result of low-resolution docking a large number of unrelated antibodies to a desired epitope on the structure of a target antigen as a prelude to de novo design.\n",
    "\n",
    "\tThe program CAN computationally design an antibody to anywhere on the target protein, but it is recommended to place the antibody at the target epitope.  It is beyond the scope of this program to determine potential epitopes for binding, however servers and programs exist to predict these. Automatic SiteConstraints can be used to further limit the design to target regions.\n",
    "\n",
    "\n",
    "3.  Model Numbering and Light Chain identification\n",
    "\n",
    "\tThe input PDB file must be renumbered to the AHo Scheme and the light chain gene must be identified.  This can be done through the [PyIgClassify Server](http://dunbrack2.fccc.edu/pyigclassify/). \n",
    "\n",
    "\tOn input into the program, Rosetta assigns our CDR clusters using the same methodology as PyIgClassify. The RosettaAntibodyDesign protocol is then driven by a set of command-line options and a set of design instructions provided as an input file that controls which CDR(s) are designed and how. Details and example command lines and instruction files are provided below.\n",
    "\n",
    "\tThe gene of the light chain should always be set on the command-line using the option `-light_chain`, these are either lamda or kappa.  PyIgClassify will identify the gene of the light chain.\n",
    "\n",
    "\tFor this tutorial, the starting antibody is renumbered for you. \n",
    "\n",
    "4.  Notes for Tutorial Shortening\n",
    "\n",
    "\tAlways set the option, `-outer_cycle_rounds` to 5 in order to run these examples quickly.  The default is 25.  We include this in our common options file that is read in by Rosetta at the start. We will only be outputting a single structure, but typical use of the protocol is with default settings of `-outer_cycle_rounds` and an `nstruct` of at least 1000, with 5000-10000 recommended for jobs that are doing a lot of grafting.  For De-novo design runs, one would want to go even higher. Note that the Docking stage increases runtime significantly as well. \n",
    "\n",
    "\tThe total number of rounds is outer_cycle_rounds * nstruct.  \n",
    "\n",
    "\t\n",
    "5.  General Notes\n",
    "\n",
    "\t\t\tsetenv PATH ${PATH}:${HOME}/rosetta_workshop/rosetta/main/source/tools\n",
    "\n",
    "\tWe will be using JSON output of the scorefile, as this is much easier to work with in python and pandas.\n",
    "\tWe use the option `-scorefile_format json`\n",
    "\n",
    "\tAll of our common options for the tutorial are in the common file that you will copy to your working directory. \n",
    "\tRosetta/PyRosetta will look for this file in your working directory or your home folder in the directory `$HOME/.rosetta/flags`.\n",
    "\tSee this page for more info on using rosetta with custom config files: <https://www.rosettacommons.org/docs/latest/rosetta_basics/running-rosetta-with-options#common-options-and-default-user-configuration>\n",
    "\n",
    "\tAll tutorials have generated output in `outputs/rabd` and their approximate time to finish on a single (core i7) processor.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Notebook setup\n",
    "!pip install pyrosettacolabsetup\n",
    "import pyrosettacolabsetup\n",
    "pyrosettacolabsetup.setup()\n",
    "print (\"Notebook is set for PyRosetta use in Colab.  Have fun!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Make sure you are in the directory with the pdb files:**\n",
    "\n",
    "`cd google_drive/My\\ Drive/student-notebooks/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import *\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import json\n",
    "import re\n",
    "\n",
    "#Functions we will be using. I like to collect any extra functions at the top of my notebook.\n",
    "def load_json_scorefile(file_path: Path, sort_by: str=\"dG_separated\") -> pandas.DataFrame:\n",
    "        \"\"\"\n",
    "        Read scorefile lines as a dataframe, sorted by total_score with Nan's correctly replaced.\n",
    "        \"\"\"\n",
    "        \n",
    "        local_lines = open(file_path, 'r').readlines()\n",
    "        decoys=[]\n",
    "        for line in local_lines:\n",
    "                o = json.loads(line.replace(\"nan\", \"NaN\"))\n",
    "                # print o[self.decoy_field_name]\n",
    "                # print repr(o)\n",
    "                decoys.append(o)\n",
    "        local_df = pandas.DataFrame.from_dict(decoys)\n",
    "        local_df = local_df.infer_objects()\n",
    "        # df.to_csv(\"debugging.csv\", sep=\",\")\n",
    "\n",
    "        local_df = local_df.sort_values(sort_by, ascending=True)\n",
    "        \n",
    "        return local_df\n",
    "\n",
    "def drop_cluster_columns(local_df: pandas.DataFrame, keep_cdrs: List[str]=None) -> pandas.DataFrame:\n",
    "        \"\"\"\n",
    "        Drop cluster columns that RAbD outputs to make it easier to work with the dataframe.\n",
    "        \"\"\"\n",
    "        to_drop = []\n",
    "        for column in local_df.columns:\n",
    "            if re.search(\"cdr_cluster\", column):\n",
    "                skip=False\n",
    "                if (keep_cdrs):\n",
    "                    for cdr in keep_cdrs:\n",
    "                        if re.search(cdr, column):\n",
    "                            skip=True\n",
    "                            break\n",
    "                if not skip:\n",
    "                    to_drop.append(column)\n",
    "        return local_df.drop(columns=to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Python\n",
    "from pyrosetta import *\n",
    "from pyrosetta.rosetta import *\n",
    "from pyrosetta.teaching import *\n",
    "\n",
    "#Core Includes\n",
    "from rosetta.protocols.rosetta_scripts import *\n",
    "from rosetta.protocols.antibody import *\n",
    "from rosetta.protocols.antibody.design import *\n",
    "from rosetta.utility import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intitlialization \n",
    "Since we are sharing the working directory with all other notebooks, instead of using the common-configuration we spoke about in the introduction, we will be using the flags file located in the inputs directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PyRosetta-4 2019 [Rosetta PyRosetta4.Release.python36.mac 2019.33+release.1e60c63beb532fd475f0f704d68d462b8af2a977 2019-08-09T15:19:57] retrieved from: http://www.pyrosetta.org\n",
      "(C) Copyright Rosetta Commons Member Institutions. Created in JHU by Sergey Lyskov and PyRosetta Team.\n",
      "\u001b[0mcore.init: \u001b[0mRosetta version: PyRosetta4.Release.python36.mac r230 2019.33+release.1e60c63beb5 1e60c63beb532fd475f0f704d68d462b8af2a977 http://www.pyrosetta.org 2019-08-09T15:19:57\n",
      "\u001b[0mcore.init: \u001b[0mcommand: PyRosetta -no_fconfig @inputs/rabd/common -database /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database\n",
      "\u001b[0mbasic.random.init_random_generator: \u001b[0m'RNG device' seed mode, using '/dev/urandom', seed=742267458 seed_offset=0 real_seed=742267458\n",
      "\u001b[0mbasic.random.init_random_generator: \u001b[0mRandomGenerator:init: Normal mode, seed=742267458 RG_type=mt19937\n"
     ]
    }
   ],
   "source": [
    "init('-no_fconfig @inputs/rabd/common')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mFinished initializing fa_standard residue type set.  Created 980 residue types\n",
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mTotal time to initialize 0.947685 seconds.\n",
      "\u001b[0mcore.import_pose.import_pose: \u001b[0mFile 'inputs/rabd/my_ab.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.io.pdb.pdb_reader: \u001b[0mParsing 993 .pdb records with unknown format to search for Rosetta-specific comments.\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 956\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYD\n"
     ]
    }
   ],
   "source": [
    "#Import a pose\n",
    "pose = pose_from_pdb(\"inputs/rabd/my_ab.pdb\")\n",
    "original_pose = pose.clone()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial\n",
    "\n",
    "\n",
    "## Tutorial A: General Design\n",
    "\t\n",
    "In many of these examples, we will use the xml interface to PyRosetta for simplicity with the AntibodyDesignMover - which is  the actual C++ application as a mover. <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/movers_pages/antibodies/AntibodyDesignMover>\n",
    "\n",
    "Lets copy the files we need first:\n",
    "\n",
    "\t\tcp ../inputs/rabd/color_cdrs.pml .\n",
    "\t\tcp ../inputs/rabd/rabd.xml .\n",
    "\n",
    "You are starting design on a new antibody that is not bound to the antigen in the crystal. This is difficult and risky, but we review how one could go about this anyway.  We start by selecting a framework.  Here, we use the trastuzumab framework as it expresses well, is thermodynamically stable with a Tm of 69.5 degrees, and has been shown repeatedly that it can tolerate CDRs of different sequence and structure.  Note that the energy of the complex is high as we are starting from a manual placement of the antibody to antigen.  If we relax the structure too much, we will fall into an energy well that is hard to escape without significant sampling. \n",
    "\n",
    "We are using an arbitrary protein at an arbitrary site for design.  The PDB of our target is 1qaw.  1qaw is an oligomer of the TRP RNA-Binding Attenuation Protein from Bacillus Stearothermophilus.  It is usually a monomer/dimer, but at its multimeric interface is a tryptophan residue by itself.  \n",
    "\n",
    "It's a beautiful protein, with a cool mechanism.  We will attempt to build an antibody to bind to two subunits to stabilize the dimeric state of the complex in the absence of TRP. Note that denovo design currently takes a large amount of processing power. Each tutorial below is more complex than the one before it.  The examples we have for this tutorial are short runs to show HOW it can be done, but more outer_cycle_rounds and nstruct would produce far better models than the ones you will see here - as we will need to sample the relative orientation of the antibody-antigen complex through docking, the CDR clusters and lengths, the internal backbone degrees of freedom of the CDRs, as well as the sequence of the CDRs and possibly the framework.  As you can tell, just the sampling problem alone is difficult. However, this will give you a basis for using RAbD on your own. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A1. Sequence Design\n",
    "\t\n",
    "Using the application is as simple as setting the `-seq_design_cdrs` option.\n",
    "This simply designs the CDRs of the heavy chain using cdr profiles if they exist for those clusters during flexible-backbone design.  If the clusters do not exist (as is the case for H3 at the moment), we use conservative design by default.  Note that InterfaceAnalyzer is run on each output decoy in the RAbD mover. Note that you can also set `light_chain` on the command line if you are only working on a single PDB through the rosetta run. \n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" light_chain=\"kappa\"/>\n",
    "            \n",
    "This will take a about a minute (50 seconds on my laptop).  Output structures and scores are in `outputs/rabd` if you wish to copy them over - these include 4 more structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mGenerating XML Schema for rosetta_scripts...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mInitializing schema validator...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mValidating input script...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mParsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<AntibodyDesignMover light_chain=\"kappa\" name=\"RAbD\" seq_design_cdrs=\"L1,L3\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "\u001b[0mcore.scoring.P_AA: \u001b[0mshapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/PairEPotential/pdb_pair_stats_fine\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/InterchainPotential/interchain_env_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/InterchainPotential/interchain_pair_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/env_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/cbeta_den.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/pair_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/cenpack_log.txt\n",
      "\u001b[0mcore.scoring.ramachandran: \u001b[0mshapovalov_lib::shap_rama_smooth_level of 4( aka highest_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/shapovalov/kappa25/all.ramaProb\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"RAbD\" of type AntibodyDesignMover\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0mParsedProtocol mover with the following movers and filters\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "<pyrosetta.rosetta.protocols.antibody.design.AntibodyDesignMover object at 0x12ee067a0>\n"
     ]
    }
   ],
   "source": [
    "rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" light_chain=\"kappa\"/>')\n",
    "rabd.apply(pose)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, for the sake of learning how to do this - how would we do this in code instead of the XML - we just need to use setters. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n"
     ]
    }
   ],
   "source": [
    "pose = original_pose.clone()\n",
    "ab_info = AntibodyInfo(pose) #We don't need to supply scheme and definition since we do so in the flags file.\n",
    "rabd2 = AntibodyDesignMover(ab_info)\n",
    "\n",
    "cdrs = vector1_protocols_antibody_CDRNameEnum()\n",
    "cdrs.append(l1)\n",
    "cdrs.append(l3)\n",
    "\n",
    "rabd2.set_seq_design_cdrs(cdrs)\n",
    "rabd2.set_light_chain(\"kappa\")\n",
    "rabd2.apply(pose)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Score the input pose using the InterfaceAnalayzerMover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mUsing explicit constructor\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mUsing interface constructor\n",
      "\u001b[0mprotocols.evaluation.ChiWellRmsdEvaluatorCreator: \u001b[0mEvaluation Creator active ...\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mInterface set residues total: 100\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mNULL scorefunction. Initialize from cmd line.\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "\u001b[0mcore.scoring.P_AA: \u001b[0mshapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/elec_cp_reps.dat\n",
      "\u001b[0mcore.scoring.elec.util: \u001b[0mRead 40 countpair representative atoms\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mshapovalov_lib_fixes_enable option is true.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mshapovalov_lib::shap_dun10_smooth_level of 1( aka lowest_smooth ) got activated.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mBinary rotamer library selected: /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mUsing Dunbrack library binary file '/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin'.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mDunbrack 2010 library took 0.250793 seconds to load from binary\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 956\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYD\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mCalculating dSASA\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mCalculating per-res dSASA data\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mincluded_nres: 975\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mFound Hbond between chains: 12 and 9\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mFound Hbond between chains: 9 and 13\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mFound Hbond between chains: 12 and 9\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mComputing delta unsat polar residues...\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/sp2_elec_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/sp2_elec_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/sp2_elec_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/sc/sc_radii.lib\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mComputing Shape Complementarity Score...\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mUpstream chain(s) numbers: 12, 13,\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mDownstream chain(s) numbers: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,\n"
     ]
    }
   ],
   "source": [
    "from rosetta.protocols.analysis import InterfaceAnalyzerMover\n",
    "\n",
    "iam = InterfaceAnalyzerMover(\"LH_ABCDEFGIJKZ\")\n",
    "iam.set_pack_separated(True)\n",
    "iam.apply(pose)\n",
    "iam.apply(original_pose)\n",
    "\n",
    "dg_term = \"dG_separated\"\n",
    "print(\"dG Diff:\", pose.scores[dg_term] - original_pose[dg_term])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Has the energy gone down after our sequence design?  The `dG_separated` is calculated by scoring the complex, separating the antigen from the antibody, repacking side-chains at the interface, and then taking the difference in score - i.e. the dG. \n",
    "    \n",
    "Lets take a look at scores from a previous run of 5 antibodies.  The scorefiles are in json format, so it will be easy to turn them into pandas Dataframes and do some cool stuff.  We'll do this often as the runtimes increase for our protocol - but all the scores in them can be accessed using the pose.scores attribute (which is PyRosetta-specific functionality.)\n",
    "\t\n",
    "Are any of these better than our input pose?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decoy</th>\n",
       "      <th>atom_pair_constraint</th>\n",
       "      <th>cdr_cluster_DIS_L1</th>\n",
       "      <th>cdr_cluster_DIS_L3</th>\n",
       "      <th>cdr_cluster_LEN_L1</th>\n",
       "      <th>cdr_cluster_LEN_L3</th>\n",
       "      <th>complex_normalized</th>\n",
       "      <th>dG_cross</th>\n",
       "      <th>dG_cross/dSASAx100</th>\n",
       "      <th>dG_separated</th>\n",
       "      <th>...</th>\n",
       "      <th>ref</th>\n",
       "      <th>sc_value</th>\n",
       "      <th>side1_normalized</th>\n",
       "      <th>side1_score</th>\n",
       "      <th>side2_normalized</th>\n",
       "      <th>side2_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>yhh_planarity</th>\n",
       "      <th>cdr_cluster_ID_L1</th>\n",
       "      <th>cdr_cluster_ID_L3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tutA1_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.244154</td>\n",
       "      <td>14.169669</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1784.542480</td>\n",
       "      <td>...</td>\n",
       "      <td>374.81054</td>\n",
       "      <td>0.491761</td>\n",
       "      <td>14.690318</td>\n",
       "      <td>837.348145</td>\n",
       "      <td>20.825792</td>\n",
       "      <td>874.683289</td>\n",
       "      <td>2432.049006</td>\n",
       "      <td>0.134353</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>tutA1_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.604385</td>\n",
       "      <td>10.078285</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.459319</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1820.213013</td>\n",
       "      <td>...</td>\n",
       "      <td>374.09441</td>\n",
       "      <td>0.596517</td>\n",
       "      <td>14.805477</td>\n",
       "      <td>858.717651</td>\n",
       "      <td>20.484030</td>\n",
       "      <td>860.329285</td>\n",
       "      <td>2397.836108</td>\n",
       "      <td>0.420196</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>tutA1_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.939879</td>\n",
       "      <td>14.834209</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1849.303833</td>\n",
       "      <td>...</td>\n",
       "      <td>372.58994</td>\n",
       "      <td>0.468819</td>\n",
       "      <td>17.130129</td>\n",
       "      <td>907.896790</td>\n",
       "      <td>20.831068</td>\n",
       "      <td>874.904846</td>\n",
       "      <td>2432.425954</td>\n",
       "      <td>0.061565</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>tutA1_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.124744</td>\n",
       "      <td>14.362237</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.667474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1934.344604</td>\n",
       "      <td>...</td>\n",
       "      <td>374.83686</td>\n",
       "      <td>0.415013</td>\n",
       "      <td>17.054668</td>\n",
       "      <td>920.952087</td>\n",
       "      <td>22.980696</td>\n",
       "      <td>965.189209</td>\n",
       "      <td>2600.787151</td>\n",
       "      <td>0.044588</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>tutA1_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.693962</td>\n",
       "      <td>12.150679</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.530113</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2854.750488</td>\n",
       "      <td>...</td>\n",
       "      <td>376.96998</td>\n",
       "      <td>0.464472</td>\n",
       "      <td>24.664927</td>\n",
       "      <td>1381.235962</td>\n",
       "      <td>33.255428</td>\n",
       "      <td>1396.727905</td>\n",
       "      <td>3441.859648</td>\n",
       "      <td>0.075642</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              decoy  atom_pair_constraint  cdr_cluster_DIS_L1  \\\n",
       "0  tutA1_my_ab_0001                   0.0           10.244154   \n",
       "3  tutA1_my_ab_0004                   0.0            9.604385   \n",
       "1  tutA1_my_ab_0002                   0.0            9.939879   \n",
       "4  tutA1_my_ab_0005                   0.0           13.124744   \n",
       "2  tutA1_my_ab_0003                   0.0            9.693962   \n",
       "\n",
       "   cdr_cluster_DIS_L3  cdr_cluster_LEN_L1  cdr_cluster_LEN_L3  \\\n",
       "0           14.169669                11.0                 9.0   \n",
       "3           10.078285                11.0                 9.0   \n",
       "1           14.834209                11.0                 9.0   \n",
       "4           14.362237                11.0                 9.0   \n",
       "2           12.150679                11.0                 9.0   \n",
       "\n",
       "   complex_normalized  dG_cross  dG_cross/dSASAx100  dG_separated  ...  \\\n",
       "0            2.494409       0.0                 0.0   1784.542480  ...   \n",
       "3            2.459319       0.0                 0.0   1820.213013  ...   \n",
       "1            2.494796       0.0                 0.0   1849.303833  ...   \n",
       "4            2.667474       0.0                 0.0   1934.344604  ...   \n",
       "2            3.530113       0.0                 0.0   2854.750488  ...   \n",
       "\n",
       "         ref  sc_value  side1_normalized  side1_score  side2_normalized  \\\n",
       "0  374.81054  0.491761         14.690318   837.348145         20.825792   \n",
       "3  374.09441  0.596517         14.805477   858.717651         20.484030   \n",
       "1  372.58994  0.468819         17.130129   907.896790         20.831068   \n",
       "4  374.83686  0.415013         17.054668   920.952087         22.980696   \n",
       "2  376.96998  0.464472         24.664927  1381.235962         33.255428   \n",
       "\n",
       "   side2_score  total_score  yhh_planarity  cdr_cluster_ID_L1  \\\n",
       "0   874.683289  2432.049006       0.134353            L1-11-1   \n",
       "3   860.329285  2397.836108       0.420196            L1-11-1   \n",
       "1   874.904846  2432.425954       0.061565            L1-11-1   \n",
       "4   965.189209  2600.787151       0.044588            L1-11-1   \n",
       "2  1396.727905  3441.859648       0.075642            L1-11-1   \n",
       "\n",
       "   cdr_cluster_ID_L3  \n",
       "0        L3-9-cis7-1  \n",
       "3        L3-9-cis7-1  \n",
       "1        L3-9-cis7-1  \n",
       "4        L3-9-cis7-1  \n",
       "2        L3-9-cis7-1  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = load_json_scorefile(\"expected_outputs/rabd/tutA1_score.sc\")\n",
    "df = drop_cluster_columns(df, keep_cdrs=[\"L1\", \"L3\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A2. Graft Design\n",
    "\n",
    "Now we will be enabling graft design AND sequence design on L1 and L3 loops.  With an nstruct (n decoys) of 5, we are doing 25 design trials total - IE 25 actual grafts. \n",
    "\n",
    "\n",
    "    <AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" graft_design_cdrs=\"L1,L3\">\n",
    "\n",
    "\n",
    "This will take a about 2-3 times as long as sequence design, as grafting a non-breaking loop takes time.  This was 738 seconds on my laptop to generate 5. Here, you will generate 1 at about 150 seconds  Ouptut structures and scores are in ../expected_outputs/rabd.\n",
    "\n",
    "\t\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Typically, we require a much higher `-outer_cycle_rounds` and number of decoys to see anything significant.  Did this improve energies in your single antibody?  How about our pre-generated ones? Load and take a look at the scorefile as a pandas DataFrame as we did above (`expected_outputs/rabd/tutA2_score.sc`). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decoy</th>\n",
       "      <th>atom_pair_constraint</th>\n",
       "      <th>cdr_cluster_DIS_L1</th>\n",
       "      <th>cdr_cluster_DIS_L3</th>\n",
       "      <th>cdr_cluster_LEN_L1</th>\n",
       "      <th>cdr_cluster_LEN_L3</th>\n",
       "      <th>complex_normalized</th>\n",
       "      <th>dG_cross</th>\n",
       "      <th>dG_cross/dSASAx100</th>\n",
       "      <th>dG_separated</th>\n",
       "      <th>...</th>\n",
       "      <th>ref</th>\n",
       "      <th>sc_value</th>\n",
       "      <th>side1_normalized</th>\n",
       "      <th>side1_score</th>\n",
       "      <th>side2_normalized</th>\n",
       "      <th>side2_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>yhh_planarity</th>\n",
       "      <th>cdr_cluster_ID_L1</th>\n",
       "      <th>cdr_cluster_ID_L3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>tutA2_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.192533</td>\n",
       "      <td>31.677544</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.450019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1751.548462</td>\n",
       "      <td>...</td>\n",
       "      <td>364.65868</td>\n",
       "      <td>0.564960</td>\n",
       "      <td>15.000569</td>\n",
       "      <td>825.031311</td>\n",
       "      <td>21.473850</td>\n",
       "      <td>837.480164</td>\n",
       "      <td>2383.868655</td>\n",
       "      <td>0.107882</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>tutA2_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.217281</td>\n",
       "      <td>24.178181</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.760204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1843.231445</td>\n",
       "      <td>...</td>\n",
       "      <td>375.17336</td>\n",
       "      <td>0.537087</td>\n",
       "      <td>13.518681</td>\n",
       "      <td>851.676880</td>\n",
       "      <td>21.286449</td>\n",
       "      <td>979.176697</td>\n",
       "      <td>2707.759790</td>\n",
       "      <td>0.221997</td>\n",
       "      <td>L1-15-1</td>\n",
       "      <td>L3-11-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tutA2_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.599037</td>\n",
       "      <td>27.562218</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3.348572</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2484.706299</td>\n",
       "      <td>...</td>\n",
       "      <td>374.05882</td>\n",
       "      <td>0.486206</td>\n",
       "      <td>22.621260</td>\n",
       "      <td>1198.926758</td>\n",
       "      <td>29.544832</td>\n",
       "      <td>1270.427734</td>\n",
       "      <td>3268.206329</td>\n",
       "      <td>0.361406</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-10-cis7,8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>tutA2_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29.971132</td>\n",
       "      <td>45.571491</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.352178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2587.653564</td>\n",
       "      <td>...</td>\n",
       "      <td>373.00912</td>\n",
       "      <td>0.620726</td>\n",
       "      <td>18.605259</td>\n",
       "      <td>1265.157593</td>\n",
       "      <td>28.719591</td>\n",
       "      <td>1292.381592</td>\n",
       "      <td>3288.486156</td>\n",
       "      <td>0.126786</td>\n",
       "      <td>L1-17-1</td>\n",
       "      <td>L3-9-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>tutA2_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.153269</td>\n",
       "      <td>17.002886</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.869139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5082.230469</td>\n",
       "      <td>...</td>\n",
       "      <td>376.47676</td>\n",
       "      <td>0.571551</td>\n",
       "      <td>42.399551</td>\n",
       "      <td>2501.573486</td>\n",
       "      <td>62.364841</td>\n",
       "      <td>2494.593750</td>\n",
       "      <td>5710.672501</td>\n",
       "      <td>0.100791</td>\n",
       "      <td>L1-10-2</td>\n",
       "      <td>L3-8-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              decoy  atom_pair_constraint  cdr_cluster_DIS_L1  \\\n",
       "4  tutA2_my_ab_0005                   0.0           11.192533   \n",
       "2  tutA2_my_ab_0003                   0.0           20.217281   \n",
       "0  tutA2_my_ab_0001                   0.0            8.599037   \n",
       "1  tutA2_my_ab_0002                   0.0           29.971132   \n",
       "3  tutA2_my_ab_0004                   0.0           12.153269   \n",
       "\n",
       "   cdr_cluster_DIS_L3  cdr_cluster_LEN_L1  cdr_cluster_LEN_L3  \\\n",
       "4           31.677544                11.0                 7.0   \n",
       "2           24.178181                15.0                11.0   \n",
       "0           27.562218                11.0                10.0   \n",
       "1           45.571491                17.0                 9.0   \n",
       "3           17.002886                10.0                 8.0   \n",
       "\n",
       "   complex_normalized  dG_cross  dG_cross/dSASAx100  dG_separated  ...  \\\n",
       "4            2.450019       0.0                 0.0   1751.548462  ...   \n",
       "2            2.760204       0.0                 0.0   1843.231445  ...   \n",
       "0            3.348572       0.0                 0.0   2484.706299  ...   \n",
       "1            3.352178       0.0                 0.0   2587.653564  ...   \n",
       "3            5.869139       0.0                 0.0   5082.230469  ...   \n",
       "\n",
       "         ref  sc_value  side1_normalized  side1_score  side2_normalized  \\\n",
       "4  364.65868  0.564960         15.000569   825.031311         21.473850   \n",
       "2  375.17336  0.537087         13.518681   851.676880         21.286449   \n",
       "0  374.05882  0.486206         22.621260  1198.926758         29.544832   \n",
       "1  373.00912  0.620726         18.605259  1265.157593         28.719591   \n",
       "3  376.47676  0.571551         42.399551  2501.573486         62.364841   \n",
       "\n",
       "   side2_score  total_score  yhh_planarity  cdr_cluster_ID_L1  \\\n",
       "4   837.480164  2383.868655       0.107882            L1-11-1   \n",
       "2   979.176697  2707.759790       0.221997            L1-15-1   \n",
       "0  1270.427734  3268.206329       0.361406            L1-11-1   \n",
       "1  1292.381592  3288.486156       0.126786            L1-17-1   \n",
       "3  2494.593750  5710.672501       0.100791            L1-10-2   \n",
       "\n",
       "   cdr_cluster_ID_L3  \n",
       "4             L3-7-1  \n",
       "2            L3-11-1  \n",
       "0     L3-10-cis7,8-1  \n",
       "1             L3-9-1  \n",
       "3             L3-8-1  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### BEGIN SOLUTION\n",
    "\n",
    "df_a2 = load_json_scorefile(\"expected_outputs/rabd/tutA2_score.sc\")\n",
    "df_a2 = drop_cluster_columns(df_a2, keep_cdrs=[\"L1\", \"L3\"])\n",
    "df_a2\n",
    "\n",
    "### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets merge these dataframes, sort by dG_separated, and see if any of our graft-design models did better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decoy</th>\n",
       "      <th>atom_pair_constraint</th>\n",
       "      <th>cdr_cluster_DIS_L1</th>\n",
       "      <th>cdr_cluster_DIS_L3</th>\n",
       "      <th>cdr_cluster_LEN_L1</th>\n",
       "      <th>cdr_cluster_LEN_L3</th>\n",
       "      <th>complex_normalized</th>\n",
       "      <th>dG_cross</th>\n",
       "      <th>dG_cross/dSASAx100</th>\n",
       "      <th>dG_separated</th>\n",
       "      <th>...</th>\n",
       "      <th>ref</th>\n",
       "      <th>sc_value</th>\n",
       "      <th>side1_normalized</th>\n",
       "      <th>side1_score</th>\n",
       "      <th>side2_normalized</th>\n",
       "      <th>side2_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>yhh_planarity</th>\n",
       "      <th>cdr_cluster_ID_L1</th>\n",
       "      <th>cdr_cluster_ID_L3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>tutA2_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.192533</td>\n",
       "      <td>31.677544</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.450019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1751.548462</td>\n",
       "      <td>...</td>\n",
       "      <td>364.65868</td>\n",
       "      <td>0.564960</td>\n",
       "      <td>15.000569</td>\n",
       "      <td>825.031311</td>\n",
       "      <td>21.473850</td>\n",
       "      <td>837.480164</td>\n",
       "      <td>2383.868655</td>\n",
       "      <td>0.107882</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tutA1_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.244154</td>\n",
       "      <td>14.169669</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1784.542480</td>\n",
       "      <td>...</td>\n",
       "      <td>374.81054</td>\n",
       "      <td>0.491761</td>\n",
       "      <td>14.690318</td>\n",
       "      <td>837.348145</td>\n",
       "      <td>20.825792</td>\n",
       "      <td>874.683289</td>\n",
       "      <td>2432.049006</td>\n",
       "      <td>0.134353</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>tutA1_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.604385</td>\n",
       "      <td>10.078285</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.459319</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1820.213013</td>\n",
       "      <td>...</td>\n",
       "      <td>374.09441</td>\n",
       "      <td>0.596517</td>\n",
       "      <td>14.805477</td>\n",
       "      <td>858.717651</td>\n",
       "      <td>20.484030</td>\n",
       "      <td>860.329285</td>\n",
       "      <td>2397.836108</td>\n",
       "      <td>0.420196</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>tutA2_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.217281</td>\n",
       "      <td>24.178181</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.760204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1843.231445</td>\n",
       "      <td>...</td>\n",
       "      <td>375.17336</td>\n",
       "      <td>0.537087</td>\n",
       "      <td>13.518681</td>\n",
       "      <td>851.676880</td>\n",
       "      <td>21.286449</td>\n",
       "      <td>979.176697</td>\n",
       "      <td>2707.759790</td>\n",
       "      <td>0.221997</td>\n",
       "      <td>L1-15-1</td>\n",
       "      <td>L3-11-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>tutA1_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.939879</td>\n",
       "      <td>14.834209</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1849.303833</td>\n",
       "      <td>...</td>\n",
       "      <td>372.58994</td>\n",
       "      <td>0.468819</td>\n",
       "      <td>17.130129</td>\n",
       "      <td>907.896790</td>\n",
       "      <td>20.831068</td>\n",
       "      <td>874.904846</td>\n",
       "      <td>2432.425954</td>\n",
       "      <td>0.061565</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>tutA1_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.124744</td>\n",
       "      <td>14.362237</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.667474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1934.344604</td>\n",
       "      <td>...</td>\n",
       "      <td>374.83686</td>\n",
       "      <td>0.415013</td>\n",
       "      <td>17.054668</td>\n",
       "      <td>920.952087</td>\n",
       "      <td>22.980696</td>\n",
       "      <td>965.189209</td>\n",
       "      <td>2600.787151</td>\n",
       "      <td>0.044588</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>tutA2_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.599037</td>\n",
       "      <td>27.562218</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3.348572</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2484.706299</td>\n",
       "      <td>...</td>\n",
       "      <td>374.05882</td>\n",
       "      <td>0.486206</td>\n",
       "      <td>22.621260</td>\n",
       "      <td>1198.926758</td>\n",
       "      <td>29.544832</td>\n",
       "      <td>1270.427734</td>\n",
       "      <td>3268.206329</td>\n",
       "      <td>0.361406</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-10-cis7,8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>tutA2_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29.971132</td>\n",
       "      <td>45.571491</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.352178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2587.653564</td>\n",
       "      <td>...</td>\n",
       "      <td>373.00912</td>\n",
       "      <td>0.620726</td>\n",
       "      <td>18.605259</td>\n",
       "      <td>1265.157593</td>\n",
       "      <td>28.719591</td>\n",
       "      <td>1292.381592</td>\n",
       "      <td>3288.486156</td>\n",
       "      <td>0.126786</td>\n",
       "      <td>L1-17-1</td>\n",
       "      <td>L3-9-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>tutA1_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.693962</td>\n",
       "      <td>12.150679</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.530113</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2854.750488</td>\n",
       "      <td>...</td>\n",
       "      <td>376.96998</td>\n",
       "      <td>0.464472</td>\n",
       "      <td>24.664927</td>\n",
       "      <td>1381.235962</td>\n",
       "      <td>33.255428</td>\n",
       "      <td>1396.727905</td>\n",
       "      <td>3441.859648</td>\n",
       "      <td>0.075642</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>tutA2_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.153269</td>\n",
       "      <td>17.002886</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.869139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5082.230469</td>\n",
       "      <td>...</td>\n",
       "      <td>376.47676</td>\n",
       "      <td>0.571551</td>\n",
       "      <td>42.399551</td>\n",
       "      <td>2501.573486</td>\n",
       "      <td>62.364841</td>\n",
       "      <td>2494.593750</td>\n",
       "      <td>5710.672501</td>\n",
       "      <td>0.100791</td>\n",
       "      <td>L1-10-2</td>\n",
       "      <td>L3-8-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              decoy  atom_pair_constraint  cdr_cluster_DIS_L1  \\\n",
       "5  tutA2_my_ab_0005                   0.0           11.192533   \n",
       "0  tutA1_my_ab_0001                   0.0           10.244154   \n",
       "1  tutA1_my_ab_0004                   0.0            9.604385   \n",
       "6  tutA2_my_ab_0003                   0.0           20.217281   \n",
       "2  tutA1_my_ab_0002                   0.0            9.939879   \n",
       "3  tutA1_my_ab_0005                   0.0           13.124744   \n",
       "7  tutA2_my_ab_0001                   0.0            8.599037   \n",
       "8  tutA2_my_ab_0002                   0.0           29.971132   \n",
       "4  tutA1_my_ab_0003                   0.0            9.693962   \n",
       "9  tutA2_my_ab_0004                   0.0           12.153269   \n",
       "\n",
       "   cdr_cluster_DIS_L3  cdr_cluster_LEN_L1  cdr_cluster_LEN_L3  \\\n",
       "5           31.677544                11.0                 7.0   \n",
       "0           14.169669                11.0                 9.0   \n",
       "1           10.078285                11.0                 9.0   \n",
       "6           24.178181                15.0                11.0   \n",
       "2           14.834209                11.0                 9.0   \n",
       "3           14.362237                11.0                 9.0   \n",
       "7           27.562218                11.0                10.0   \n",
       "8           45.571491                17.0                 9.0   \n",
       "4           12.150679                11.0                 9.0   \n",
       "9           17.002886                10.0                 8.0   \n",
       "\n",
       "   complex_normalized  dG_cross  dG_cross/dSASAx100  dG_separated  ...  \\\n",
       "5            2.450019       0.0                 0.0   1751.548462  ...   \n",
       "0            2.494409       0.0                 0.0   1784.542480  ...   \n",
       "1            2.459319       0.0                 0.0   1820.213013  ...   \n",
       "6            2.760204       0.0                 0.0   1843.231445  ...   \n",
       "2            2.494796       0.0                 0.0   1849.303833  ...   \n",
       "3            2.667474       0.0                 0.0   1934.344604  ...   \n",
       "7            3.348572       0.0                 0.0   2484.706299  ...   \n",
       "8            3.352178       0.0                 0.0   2587.653564  ...   \n",
       "4            3.530113       0.0                 0.0   2854.750488  ...   \n",
       "9            5.869139       0.0                 0.0   5082.230469  ...   \n",
       "\n",
       "         ref  sc_value  side1_normalized  side1_score  side2_normalized  \\\n",
       "5  364.65868  0.564960         15.000569   825.031311         21.473850   \n",
       "0  374.81054  0.491761         14.690318   837.348145         20.825792   \n",
       "1  374.09441  0.596517         14.805477   858.717651         20.484030   \n",
       "6  375.17336  0.537087         13.518681   851.676880         21.286449   \n",
       "2  372.58994  0.468819         17.130129   907.896790         20.831068   \n",
       "3  374.83686  0.415013         17.054668   920.952087         22.980696   \n",
       "7  374.05882  0.486206         22.621260  1198.926758         29.544832   \n",
       "8  373.00912  0.620726         18.605259  1265.157593         28.719591   \n",
       "4  376.96998  0.464472         24.664927  1381.235962         33.255428   \n",
       "9  376.47676  0.571551         42.399551  2501.573486         62.364841   \n",
       "\n",
       "   side2_score  total_score  yhh_planarity  cdr_cluster_ID_L1  \\\n",
       "5   837.480164  2383.868655       0.107882            L1-11-1   \n",
       "0   874.683289  2432.049006       0.134353            L1-11-1   \n",
       "1   860.329285  2397.836108       0.420196            L1-11-1   \n",
       "6   979.176697  2707.759790       0.221997            L1-15-1   \n",
       "2   874.904846  2432.425954       0.061565            L1-11-1   \n",
       "3   965.189209  2600.787151       0.044588            L1-11-1   \n",
       "7  1270.427734  3268.206329       0.361406            L1-11-1   \n",
       "8  1292.381592  3288.486156       0.126786            L1-17-1   \n",
       "4  1396.727905  3441.859648       0.075642            L1-11-1   \n",
       "9  2494.593750  5710.672501       0.100791            L1-10-2   \n",
       "\n",
       "   cdr_cluster_ID_L3  \n",
       "5             L3-7-1  \n",
       "0        L3-9-cis7-1  \n",
       "1        L3-9-cis7-1  \n",
       "6            L3-11-1  \n",
       "2        L3-9-cis7-1  \n",
       "3        L3-9-cis7-1  \n",
       "7     L3-10-cis7,8-1  \n",
       "8             L3-9-1  \n",
       "4        L3-9-cis7-1  \n",
       "9             L3-8-1  \n",
       "\n",
       "[10 rows x 48 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tut_a12 = pandas.concat([df, df_a2], ignore_index=True).sort_values(\"dG_separated\", ascending=True)\n",
    "df_tut_a12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take a look at the lowest (dG) scoring pose in pymol - do you see any difference in L1 and L3 loops there?  Do they make better contact than what we had before?\n",
    "\n",
    "\tLets take a look in pymol.   \n",
    "\n",
    "\t\t\tpymol inputs/rabd/my_ab.pdb inputs/rabd/tutA2_* \n",
    "\t\t\t@color_cdrs.pml\n",
    "\t\t\tcenter full_epitope\n",
    "\n",
    " How different are the L1 and L3 loops? Have any changed length?\n",
    "\n",
    " Lets take a look at the clusters in our dataframe.  Have they changed from the native?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"L1\", original_pose.scores[\"cdr_cluster_ID_L1\"])\n",
    "print(\"L3\", original_pose.scores[\"cdr_cluster_ID_L3\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A3.  Basic De-novo run\n",
    "\n",
    "Here, we want to do a denovo-run (without docking), starting with random CDRs grafted in - instead of whatever we have in the antibody to start with (only for the CDRs that are actually undergoing graft-design).  This is useful, as we start the design with very high energy and work our way down.  Note that since this is an entirely new interface for our model protein, this interface is already at a very high energy - and so this is less needed, but it should be noted how to do this. (139 seconds on my laptop). Do this below as you have done in other tutorials - either through code or XML.\n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" graft_design_cdrs=\"L1,L3\" seq_design_cdrs=\"L1,L3\" \n",
    "\t\t\t                                                                  random_start=\"1\"/>  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### BEGIN SOLUTION\n",
    "pose = original_pose.clone()\n",
    "rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" graft_design_cdrs=\"L1,L3\" seq_design_cdrs=\"L1,L3\" random_start=\"1\"/>')\n",
    "rabd.apply(pose)\n",
    "\n",
    "## OR (REUSE code from above)\n",
    "pose = original_pose.clone()\n",
    "rabd2.set_seq_design_cdrs(cdrs)\n",
    "rabd2.set_graft_design_cdrs(cdrs)\n",
    "rabd2.set_random_start(True)\n",
    "rabd2.set_light_chain(\"kappa\")\n",
    "rabd2.apply(pose)\n",
    "\n",
    "### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would starting from a random CDR help anywhere?  Perhaps if you want an entirely new cluster or length to break a patent or remove some off target effects?  We will use it below to start de novo design with docking."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. RosettaScripts RAbD Framework Components\n",
    "\n",
    "This tutorial will give you some exprience with your own XML antibody design protocol using the RosettaAntibdyDesign components.  We will take the light chain CDRs from a malaria antibody and graft them into our antibody.  In the tutorial we are interested in stabilizing the grafted CDRs in relation to the whole antibody, instead of interface design to an antigen.  \n",
    "\n",
    "We will graft the CDRs in, minimize the structure with CDR dihedral constraints to not purturb the CDRs too much, and then design the framework around the CDRs while designing the CDRs and neighbors.  The result should be mutations that better accomodate our new CDRs.  This can be useful for humanizing potential antibodies or framework switching, where we want the binding properties of certain CDRs, but the stability or immunological profile of a different framework.  \n",
    "\t\n",
    "\n",
    "#### 1. Copy the Files\n",
    "\n",
    "        cp ../input_files/ab_design_components.xml .\n",
    "\t\tcp ../input_files/malaria_cdrs.pdb .\n",
    "\t\n",
    "\t\t\n",
    "Take a look at the xml.  We are using the _AntibodyCDRGrafter_ to do the grafting of our CDRs.  We then add constraints using _CDRDihderalConstraintMover_s for each CDR.  Next, we do a round of pack/min/pack using the _RestrictToCDRsAndNeighborsOperation_ and the CDRResidueSelector.  This task operation controls what we pack and design.  It first limits packing and design to only the CDRs and its neighbors.  By specifying the `design_framework=1` option we allow the neighbor framework residues to design, while the CDRs and antigen neighbors will only repack.  If we wanted to disable antigen repacking, we would pass the _DisableAntibodyRegionOperation_ task operation.  Using this, we can specify any antibody region as `antibody_region`, `cdr_region`, or `antigen_region` and we can disable just design or both packing and design.  \n",
    "\t\t\n",
    "\n",
    "These task operations allow us to chisel exactly what we want to design in antibody, sans a residue-specific resfile (though we could combine these with one of them!).  All of these tools are available in-code.  If you've done the design workshop, you will know how to use them here. Checkout `rosetta.protocols.antibody.task_operations` for a list of them.\n",
    "\t\t\n",
    "Finally, we use the new SimpleMetric system to obtain our final sequence of the CDRs to compare to our native antibody as well as pymol selections of our CDRs - which you have been introduced to in the previous tutorial. \n",
    "\n",
    "If you want a challenge - try to set these up in-code without RosettaScripts.  It can be tricky - which is why I made PyRosetta finally work optionally with RosettaScripts.  Its good to know how to use both.  \n",
    "\n",
    "- <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/SimpleMetrics/SimpleMetrics>\n",
    "\n",
    "- More Documentation is available here:\n",
    "\n",
    " - <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/RosettaScripts>\n",
    "\n",
    " - <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/TaskOperations/TaskOperations-RosettaScripts#antibody-and-cdr-specific-operations>\n",
    "\n",
    " - <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/Movers-RosettaScripts#antibody-modeling-and-design-movers>\n",
    "\t\n",
    "#### 2. Run the protocol or copy the output (357 seconds).  \n",
    "\n",
    "\trosetta_scripts.linuxgccrelease -parser:protocol ab_design_components.xml \\\n",
    "\t\t\t    -input_ab_scheme AHO_Scheme -cdr_definition North \\\n",
    "\t\t\t    -s my_ab.pdb -out:prefix tutA4_ \\\n",
    "\t\t\t    -nstruct 5 -score:weights ref2015_cart -check_cdr_chainbreaks false\n",
    "\n",
    "\n",
    "#### 3. Look at the score file as you have before.  Are the sequences different between what we started with?  How about the interaction energies?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [Side Chain Conformations and Dunbrack Energies](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.01-Side-Chain-Conformations-and-Dunbrack-Energies.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [Protein Design with a Resfile and FastRelax](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.03-Design-with-a-resfile-and-relax.ipynb) ><p><a href=\"https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.02-Packing-design-and-regional-relax.ipynb\"><img align=\"left\" src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open in Colab\" title=\"Open in Google Colaboratory\"></a>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
