{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NOTEBOOK_HEADER-->\n",
    "*This notebook contains material from [PyRosetta](https://RosettaCommons.github.io/PyRosetta.notebooks);\n",
    "content is available [on Github](https://github.com/RosettaCommons/PyRosetta.notebooks.git).*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [RosettaAntibody Framework](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.01-RosettaAntibody-Framework-and-SimpleMetrics.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [RosettaCarbohydrates](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/13.00-RosettaCarbohydrates-Working-with-Glycans.ipynb) ><p><a href=\"https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.02-RosettaAntibodyDesign-RAbD.ipynb\"><img align=\"left\" src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open in Colab\" title=\"Open in Google Colaboratory\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RosettaAntibodyDesign\n",
    "## Notes\n",
    "This tutorial will walk you through how to use `RosettaAntibodyDesign` in PyRosetta.  You should also go through the parellel distribution workshop as you will most likely need to create many decoys for some of these design tasks. Note that we are using the XML interface to the code here for simplicity (and because I had a C++ workshop I am converting - truth be told).  The code-level interface is as robust as the XML - but will require more knowledge use.  You are welcome to play around with it - all functions have descriptions and all options are possible to change through code.\n",
    "\n",
    "Grab a coffee, take a breath, and lets learn how to design some antibodies!\n",
    "\n",
    "## Citation\n",
    "\n",
    "\n",
    "[Rosetta Antibody Design (RAbD): A General Framework for Computational Antibody Design, PLOS Computational Biology, 4/27/2018](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006112)\n",
    "\n",
    "Jared Adolf-Bryfogle, Oleks Kalyuzhniy, Michael Kubitz, Brian D. Weitzner, Xiaozhen Hu, Yumiko Adachi, William R. Schief, Roland L. Dunbrack Jr.\n",
    "\n",
    "## Manual\n",
    "The full RAbD manual can be found here: https://www.rosettacommons.org/docs/latest/application_documentation/antibody/RosettaAntibodyDesign\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "__RosettaAntibodyDesign (RAbD)__ is a generalized framework for the design of antibodies, in which a user can easily tailor the run to their project needs.  __The algorithm is meant to sample the diverse sequence, structure, and binding space of an antibody-antigen complex.__  An app is available, and all components can be used within RosettaScripts for easy scripting of antibody design and incorporation into other Rosetta protocols.\n",
    "\n",
    "The framework is based on rigorous bioinformatic analysis and rooted very much on our [recent clustering](https://www.ncbi.nlm.nih.gov/pubmed/21035459) of antibody CDR regions.  It uses the __North/Dunbrack CDR definition__ as outlined in the North/Dunbrack clustering paper. A new clustering paper will be out in the next year, and this new analysis will be incorporated into RAbD. \n",
    "\n",
    "The supplemental methods section of the published paper has all details of the RosettaAntibodyDesign method.  This manual serves to get you started running RAbD in typical use fashions. \n",
    "\n",
    "\n",
    "# Algorithm\n",
    "  \n",
    "Broadly, the RAbD protocol consists of alternating outer and inner Monte Carlo cycles. Each outer cycle consists of randomly choosing a CDR (L1, L2, etc.) from those CDRs set to design, randomly choosing a cluster and then a structure from that cluster from the database according to the input instructions, and optionally grafting that CDR's structure onto the antibody framework in place of the existing CDR (__GraftDesign__). The program then performs N rounds of the inner cycle, consisting of sequence design (__SeqDesign__) using cluster-based sequence profiles and structural constraints, energy minimization, and optional docking. Each inner cycle structurally optimizes the backbone and repacks side chains of the CDR chosen in the outer cycle as well as optional neighbors in order to optimize interactions of the CDR with the antigen and other CDRs. \n",
    "\n",
    "__Backbone dihedral angle (CircularHarmonic) constraints__ derived from the cluster data are applied to each CDR to limit deleterious structural perturbations. Amino acid changes are typically sampled from __profiles derived for each CDR cluster in PyIgClassify__. Conservative amino acid substitutions (according to the BLOSUM62 substitution matrix) may be performed when too few sequences are available to produce a profile (e.g., for H3). After each inner cycle is completed, the new sequence and structure are accepted according to the Metropolis Monte Carlo criterion. After N rounds within the inner cycle, the program returns to the outer cycle, at which point the energy of the resulting design is compared to the previous design in the outer cycle. The new design is accepted or rejected according to the Monte Carlo criterion.\n",
    "\n",
    "If optimizing the antibody-antigen orientation during the design (dock), SiteConstraints are automatically used to keep the CDRs (paratope) facing the antigen surface.  These are termed __ParatopeSiteConstraints__.   Optionally, one can enable constraints that keep the paratope of the antibody around a target epitope (antigen binding site).  These are called __ParatopeEpitopeSiteConstraints__ as the constraints are between the paratope and the epitope. The epitope is automatically determined as the interface residues around the paratope on input into the program, however, any residue(s) can be set as the epitope to limit unwanted movement and sampling of the antibody.  See the examples and options below. \n",
    "\n",
    "More detail on the algorithm can be found in the published paper. \n",
    "\n",
    "\n",
    "# General Setup and Inputs\n",
    "\n",
    "1.  Antibody Design Database\n",
    "\n",
    "\tThis app requires the Rosetta Antibody Design Database.  A database of antibodies from the original North Clustering paper is included in Rosetta and is used as the default .  An updated database (which is currently updated bi-yearly) can be downloaded here: <http://dunbrack2.fccc.edu/PyIgClassify/>.  \n",
    "\n",
    "\tFor C++, It should be placed in `Rosetta/main/database/sampling/antibodies/`. For PyRosetta, use the cmd-line option `antibody_database` and set it to the full path of the downloaded database within the `init()` function as you have done in the past. It is recommended to use this up-to-date database for production runs. For this tutorial, we will use the database within Rosetta. \n",
    "\n",
    "\n",
    "2.  Starting Structure\n",
    "\n",
    "\tThe protocol begins with the three-dimensional structure of an antibody-antigen complex. Designs should start with an antibody bound to a target antigen (however optimizing just the antibody without the complex is also possible).  Camelid antibodies are fully supported.  This structure may be an experimental structure of an existing antibody in complex with its antigen, a predicted structure of an existing antibody docked computationally to its antigen, or even the best scoring result of low-resolution docking a large number of unrelated antibodies to a desired epitope on the structure of a target antigen as a prelude to de novo design.\n",
    "\n",
    "\tThe program CAN computationally design an antibody to anywhere on the target protein, but it is recommended to place the antibody at the target epitope.  It is beyond the scope of this program to determine potential epitopes for binding, however servers and programs exist to predict these. Automatic SiteConstraints can be used to further limit the design to target regions.\n",
    "\n",
    "\n",
    "3.  Model Numbering and Light Chain identification\n",
    "\n",
    "\tThe input PDB file must be renumbered to the AHo Scheme and the light chain gene must be identified.  This can be done through the [PyIgClassify Server](http://dunbrack2.fccc.edu/pyigclassify/). \n",
    "\n",
    "\tOn input into the program, Rosetta assigns our CDR clusters using the same methodology as PyIgClassify. The RosettaAntibodyDesign protocol is then driven by a set of command-line options and a set of design instructions provided as an input file that controls which CDR(s) are designed and how. Details and example command lines and instruction files are provided below.\n",
    "\n",
    "\tThe gene of the light chain should always be set on the command-line using the option `-light_chain`, these are either lamda or kappa.  PyIgClassify will identify the gene of the light chain.\n",
    "\n",
    "\tFor this tutorial, the starting antibody is renumbered for you. \n",
    "\n",
    "4.  Notes for Tutorial Shortening\n",
    "\n",
    "\tAlways set the option, `-outer_cycle_rounds` to 5 in order to run these examples quickly.  The default is 25.  We include this in our common options file that is read in by Rosetta at the start. We will only be outputting a single structure, but typical use of the protocol is with default settings of `-outer_cycle_rounds` and an `nstruct` of at least 1000, with 5000-10000 recommended for jobs that are doing a lot of grafting.  For De-novo design runs, one would want to go even higher. Note that the Docking stage increases runtime significantly as well. \n",
    "\n",
    "\tThe total number of rounds is outer_cycle_rounds * nstruct.  \n",
    "\n",
    "\t\n",
    "5.  General Notes\n",
    "\n",
    "\t\t\tsetenv PATH ${PATH}:${HOME}/rosetta_workshop/rosetta/main/source/tools\n",
    "\n",
    "\tWe will be using JSON output of the scorefile, as this is much easier to work with in python and pandas.\n",
    "\tWe use the option `-scorefile_format json`\n",
    "\n",
    "\tAll of our common options for the tutorial are in the common file that you will copy to your working directory. \n",
    "\tRosetta/PyRosetta will look for this file in your working directory or your home folder in the directory `$HOME/.rosetta/flags`.\n",
    "\tSee this page for more info on using rosetta with custom config files: <https://www.rosettacommons.org/docs/latest/rosetta_basics/running-rosetta-with-options#common-options-and-default-user-configuration>\n",
    "\n",
    "\tAll tutorials have generated output in `outputs/rabd` and their approximate time to finish on a single (core i7) processor.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install pyrosettacolabsetup\n",
    "import pyrosettacolabsetup; pyrosettacolabsetup.install_pyrosetta()\n",
    "import pyrosetta; pyrosetta.init()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Make sure you are in the directory with the pdb files:**\n",
    "\n",
    "`cd google_drive/My\\ Drive/student-notebooks/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import *\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import json\n",
    "import re\n",
    "\n",
    "#Functions we will be using. I like to collect any extra functions at the top of my notebook.\n",
    "def load_json_scorefile(file_path: Path, sort_by: str=\"dG_separated\") -> pandas.DataFrame:\n",
    "        \"\"\"\n",
    "        Read scorefile lines as a dataframe, sorted by total_score with Nan's correctly replaced.\n",
    "        \"\"\"\n",
    "        \n",
    "        local_lines = open(file_path, 'r').readlines()\n",
    "        decoys=[]\n",
    "        for line in local_lines:\n",
    "                o = json.loads(line.replace(\"nan\", \"NaN\"))\n",
    "                # print o[self.decoy_field_name]\n",
    "                # print repr(o)\n",
    "                decoys.append(o)\n",
    "        local_df = pandas.DataFrame.from_dict(decoys)\n",
    "        local_df = local_df.infer_objects()\n",
    "        # df.to_csv(\"debugging.csv\", sep=\",\")\n",
    "\n",
    "        local_df = local_df.sort_values(sort_by, ascending=True)\n",
    "        \n",
    "        return local_df\n",
    "\n",
    "def drop_cluster_columns(local_df: pandas.DataFrame, keep_cdrs: List[str]=None) -> pandas.DataFrame:\n",
    "        \"\"\"\n",
    "        Drop cluster columns that RAbD outputs to make it easier to work with the dataframe.\n",
    "        \"\"\"\n",
    "        to_drop = []\n",
    "        for column in local_df.columns:\n",
    "            if re.search(\"cdr_cluster\", column):\n",
    "                skip=False\n",
    "                if (keep_cdrs):\n",
    "                    for cdr in keep_cdrs:\n",
    "                        if re.search(cdr, column):\n",
    "                            skip=True\n",
    "                            break\n",
    "                if not skip:\n",
    "                    to_drop.append(column)\n",
    "        return local_df.drop(columns=to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:8: UserWarning: Import of 'rosetta' as a top-level module is deprecated and may be removed in 2018, import via 'pyrosetta.rosetta'.\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "#Python\n",
    "from pyrosetta import *\n",
    "from pyrosetta.rosetta import *\n",
    "from pyrosetta.teaching import *\n",
    "import os\n",
    "\n",
    "#Core Includes\n",
    "from rosetta.protocols.rosetta_scripts import *\n",
    "from rosetta.protocols.antibody import *\n",
    "from rosetta.protocols.antibody.design import *\n",
    "from rosetta.utility import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intitlialization \n",
    "Since we are sharing the working directory with all other notebooks, instead of using the common-configuration we spoke about in the introduction, we will be using the flags file located in the inputs directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PyRosetta-4 2019 [Rosetta PyRosetta4.Release.python36.mac 2019.39+release.93456a567a8125cafdf7f8cb44400bc20b570d81 2019-09-26T14:24:44] retrieved from: http://www.pyrosetta.org\n",
      "(C) Copyright Rosetta Commons Member Institutions. Created in JHU by Sergey Lyskov and PyRosetta Team.\n",
      "\u001b[0mcore.init: \u001b[0mRosetta version: PyRosetta4.Release.python36.mac r233 2019.39+release.93456a567a8 93456a567a8125cafdf7f8cb44400bc20b570d81 http://www.pyrosetta.org 2019-09-26T14:24:44\n",
      "\u001b[0mcore.init: \u001b[0mcommand: PyRosetta -no_fconfig @inputs/rabd/common -database /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database\n",
      "\u001b[0mbasic.random.init_random_generator: \u001b[0m'RNG device' seed mode, using '/dev/urandom', seed=-443789040 seed_offset=0 real_seed=-443789040\n",
      "\u001b[0mbasic.random.init_random_generator: \u001b[0mRandomGenerator:init: Normal mode, seed=-443789040 RG_type=mt19937\n"
     ]
    }
   ],
   "source": [
    "init('-no_fconfig @inputs/rabd/common')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mFinished initializing fa_standard residue type set.  Created 980 residue types\n",
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mTotal time to initialize 0.889319 seconds.\n",
      "\u001b[0mcore.import_pose.import_pose: \u001b[0mFile 'inputs/rabd/my_ab.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.io.pdb.pdb_reader: \u001b[0mParsing 993 .pdb records with unknown format to search for Rosetta-specific comments.\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 956\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYD\n"
     ]
    }
   ],
   "source": [
    "#Import a pose\n",
    "pose = pose_from_pdb(\"inputs/rabd/my_ab.pdb\")\n",
    "original_pose = pose.clone()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial\n",
    "\n",
    "\n",
    "## Tutorial A: General Design\n",
    "\t\n",
    "In many of these examples, we will use the xml interface to PyRosetta for simplicity with the AntibodyDesignMover - which is  the actual C++ application as a mover. <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/movers_pages/antibodies/AntibodyDesignMover>\n",
    "\n",
    "Lets copy the files we need first:\n",
    "\n",
    "\t\tcp ../inputs/rabd/color_cdrs.pml .\n",
    "\t\tcp ../inputs/rabd/rabd.xml .\n",
    "\n",
    "You are starting design on a new antibody that is not bound to the antigen in the crystal. This is difficult and risky, but we review how one could go about this anyway.  We start by selecting a framework.  Here, we use the trastuzumab framework as it expresses well, is thermodynamically stable with a Tm of 69.5 degrees, and has been shown repeatedly that it can tolerate CDRs of different sequence and structure.  Note that the energy of the complex is high as we are starting from a manual placement of the antibody to antigen.  If we relax the structure too much, we will fall into an energy well that is hard to escape without significant sampling. \n",
    "\n",
    "We are using an arbitrary protein at an arbitrary site for design.  The PDB of our target is 1qaw.  1qaw is an oligomer of the TRP RNA-Binding Attenuation Protein from Bacillus Stearothermophilus.  It is usually a monomer/dimer, but at its multimeric interface is a tryptophan residue by itself.  \n",
    "\n",
    "It's a beautiful protein, with a cool mechanism.  We will attempt to build an antibody to bind to two subunits to stabilize the dimeric state of the complex in the absence of TRP. Note that denovo design currently takes a large amount of processing power. Each tutorial below is more complex than the one before it.  The examples we have for this tutorial are short runs to show HOW it can be done, but more outer_cycle_rounds and nstruct would produce far better models than the ones you will see here - as we will need to sample the relative orientation of the antibody-antigen complex through docking, the CDR clusters and lengths, the internal backbone degrees of freedom of the CDRs, as well as the sequence of the CDRs and possibly the framework.  As you can tell, just the sampling problem alone is difficult. However, this will give you a basis for using RAbD on your own. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut A1. Sequence Design\n",
    "\t\n",
    "Using the application is as simple as setting the `-seq_design_cdrs` option.\n",
    "This simply designs the CDRs of the heavy chain using cdr profiles if they exist for those clusters during flexible-backbone design.  If the clusters do not exist (as is the case for H3 at the moment), we use conservative design by default.  Note that InterfaceAnalyzer is run on each output decoy in the RAbD mover. Note that you can also set `light_chain` on the command line if you are only working on a single PDB through the rosetta run. \n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" light_chain=\"kappa\"/>\n",
    "            \n",
    "This will take a about a minute (50 seconds on my laptop).  Output structures and scores are in `outputs/rabd` if you wish to copy them over - these include 4 more structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mGenerating XML Schema for rosetta_scripts...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mInitializing schema validator...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mValidating input script...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mParsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<MOVERS>\n",
      "\t\t<AntibodyDesignMover light_chain=\"kappa\" name=\"RAbD\" seq_design_cdrs=\"L1,L3\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS/>\n",
      "</ROSETTASCRIPTS>\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "\u001b[0mcore.scoring.P_AA: \u001b[0mshapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/PairEPotential/pdb_pair_stats_fine\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/InterchainPotential/interchain_env_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/InterchainPotential/interchain_pair_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/env_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/cbeta_den.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/pair_log.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/EnvPairPotential/cenpack_log.txt\n",
      "\u001b[0mcore.scoring.ramachandran: \u001b[0mshapovalov_lib::shap_rama_smooth_level of 4( aka highest_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/shapovalov/kappa25/all.ramaProb\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"RAbD\" of type AntibodyDesignMover\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0mParsedProtocol mover with the following movers and filters\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "<pyrosetta.rosetta.protocols.antibody.design.AntibodyDesignMover object at 0x12ee067a0>\n"
     ]
    }
   ],
   "source": [
    "rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" light_chain=\"kappa\"/>')\n",
    "if not os.getenv(\"DEBUG\"):\n",
    "    rabd.apply(pose)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, for the sake of learning how to do this - how would we do this in code instead of the XML - we just need to use setters. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n"
     ]
    }
   ],
   "source": [
    "pose = original_pose.clone()\n",
    "rabd2 = AntibodyDesignMover()\n",
    "\n",
    "cdrs = vector1_protocols_antibody_CDRNameEnum()\n",
    "cdrs.append(l1)\n",
    "cdrs.append(l3)\n",
    "\n",
    "rabd2.set_seq_design_cdrs(cdrs)\n",
    "rabd2.set_light_chain(\"kappa\")\n",
    "if not os.getenv(\"DEBUG\"):\n",
    "    rabd2.apply(pose)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Score the input pose using the InterfaceAnalayzerMover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mUsing explicit constructor\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mUsing interface constructor\n",
      "\u001b[0mprotocols.evaluation.ChiWellRmsdEvaluatorCreator: \u001b[0mEvaluation Creator active ...\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mInterface set residues total: 100\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mNULL scorefunction. Initialize from cmd line.\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "\u001b[0mcore.scoring.P_AA: \u001b[0mshapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/elec_cp_reps.dat\n",
      "\u001b[0mcore.scoring.elec.util: \u001b[0mRead 40 countpair representative atoms\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mshapovalov_lib_fixes_enable option is true.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mshapovalov_lib::shap_dun10_smooth_level of 1( aka lowest_smooth ) got activated.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mBinary rotamer library selected: /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mUsing Dunbrack library binary file '/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin'.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mDunbrack 2010 library took 0.250793 seconds to load from binary\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 956\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 956 CYD\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mCalculating dSASA\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mCalculating per-res dSASA data\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mincluded_nres: 975\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mFound Hbond between chains: 12 and 9\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mFound Hbond between chains: 9 and 13\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mFound Hbond between chains: 12 and 9\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mComputing delta unsat polar residues...\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/sp2_elec_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/sp2_elec_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/sp2_elec_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/sc/sc_radii.lib\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mComputing Shape Complementarity Score...\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mUpstream chain(s) numbers: 12, 13,\n",
      "\u001b[0mprotocols.analysis.InterfaceAnalyzerMover: \u001b[0mDownstream chain(s) numbers: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,\n"
     ]
    }
   ],
   "source": [
    "from rosetta.protocols.analysis import InterfaceAnalyzerMover\n",
    "\n",
    "if not os.getenv(\"DEBUG\"):\n",
    "    iam = InterfaceAnalyzerMover(\"LH_ABCDEFGIJKZ\")\n",
    "    iam.set_pack_separated(True)\n",
    "    iam.apply(pose)\n",
    "    iam.apply(original_pose)\n",
    "\n",
    "    dg_term = \"dG_separated\"\n",
    "    print(\"dG Diff:\", pose.scores[dg_term] - original_pose[dg_term])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Has the energy gone down after our sequence design?  The `dG_separated` is calculated by scoring the complex, separating the antigen from the antibody, repacking side-chains at the interface, and then taking the difference in score - i.e. the dG. \n",
    "    \n",
    "Lets take a look at scores from a previous run of 5 antibodies.  The scorefiles are in json format, so it will be easy to turn them into pandas Dataframes and do some cool stuff.  We'll do this often as the runtimes increase for our protocol - but all the scores in them can be accessed using the pose.scores attribute (which is PyRosetta-specific functionality.)\n",
    "\t\n",
    "Are any of these better than our input pose?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decoy</th>\n",
       "      <th>atom_pair_constraint</th>\n",
       "      <th>cdr_cluster_DIS_L1</th>\n",
       "      <th>cdr_cluster_DIS_L3</th>\n",
       "      <th>cdr_cluster_LEN_L1</th>\n",
       "      <th>cdr_cluster_LEN_L3</th>\n",
       "      <th>complex_normalized</th>\n",
       "      <th>dG_cross</th>\n",
       "      <th>dG_cross/dSASAx100</th>\n",
       "      <th>dG_separated</th>\n",
       "      <th>...</th>\n",
       "      <th>ref</th>\n",
       "      <th>sc_value</th>\n",
       "      <th>side1_normalized</th>\n",
       "      <th>side1_score</th>\n",
       "      <th>side2_normalized</th>\n",
       "      <th>side2_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>yhh_planarity</th>\n",
       "      <th>cdr_cluster_ID_L1</th>\n",
       "      <th>cdr_cluster_ID_L3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tutA1_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.244154</td>\n",
       "      <td>14.169669</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1784.542480</td>\n",
       "      <td>...</td>\n",
       "      <td>374.81054</td>\n",
       "      <td>0.491761</td>\n",
       "      <td>14.690318</td>\n",
       "      <td>837.348145</td>\n",
       "      <td>20.825792</td>\n",
       "      <td>874.683289</td>\n",
       "      <td>2432.049006</td>\n",
       "      <td>0.134353</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>tutA1_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.604385</td>\n",
       "      <td>10.078285</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.459319</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1820.213013</td>\n",
       "      <td>...</td>\n",
       "      <td>374.09441</td>\n",
       "      <td>0.596517</td>\n",
       "      <td>14.805477</td>\n",
       "      <td>858.717651</td>\n",
       "      <td>20.484030</td>\n",
       "      <td>860.329285</td>\n",
       "      <td>2397.836108</td>\n",
       "      <td>0.420196</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>tutA1_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.939879</td>\n",
       "      <td>14.834209</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1849.303833</td>\n",
       "      <td>...</td>\n",
       "      <td>372.58994</td>\n",
       "      <td>0.468819</td>\n",
       "      <td>17.130129</td>\n",
       "      <td>907.896790</td>\n",
       "      <td>20.831068</td>\n",
       "      <td>874.904846</td>\n",
       "      <td>2432.425954</td>\n",
       "      <td>0.061565</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>tutA1_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.124744</td>\n",
       "      <td>14.362237</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.667474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1934.344604</td>\n",
       "      <td>...</td>\n",
       "      <td>374.83686</td>\n",
       "      <td>0.415013</td>\n",
       "      <td>17.054668</td>\n",
       "      <td>920.952087</td>\n",
       "      <td>22.980696</td>\n",
       "      <td>965.189209</td>\n",
       "      <td>2600.787151</td>\n",
       "      <td>0.044588</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>tutA1_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.693962</td>\n",
       "      <td>12.150679</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.530113</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2854.750488</td>\n",
       "      <td>...</td>\n",
       "      <td>376.96998</td>\n",
       "      <td>0.464472</td>\n",
       "      <td>24.664927</td>\n",
       "      <td>1381.235962</td>\n",
       "      <td>33.255428</td>\n",
       "      <td>1396.727905</td>\n",
       "      <td>3441.859648</td>\n",
       "      <td>0.075642</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              decoy  atom_pair_constraint  cdr_cluster_DIS_L1  \\\n",
       "0  tutA1_my_ab_0001                   0.0           10.244154   \n",
       "3  tutA1_my_ab_0004                   0.0            9.604385   \n",
       "1  tutA1_my_ab_0002                   0.0            9.939879   \n",
       "4  tutA1_my_ab_0005                   0.0           13.124744   \n",
       "2  tutA1_my_ab_0003                   0.0            9.693962   \n",
       "\n",
       "   cdr_cluster_DIS_L3  cdr_cluster_LEN_L1  cdr_cluster_LEN_L3  \\\n",
       "0           14.169669                11.0                 9.0   \n",
       "3           10.078285                11.0                 9.0   \n",
       "1           14.834209                11.0                 9.0   \n",
       "4           14.362237                11.0                 9.0   \n",
       "2           12.150679                11.0                 9.0   \n",
       "\n",
       "   complex_normalized  dG_cross  dG_cross/dSASAx100  dG_separated  ...  \\\n",
       "0            2.494409       0.0                 0.0   1784.542480  ...   \n",
       "3            2.459319       0.0                 0.0   1820.213013  ...   \n",
       "1            2.494796       0.0                 0.0   1849.303833  ...   \n",
       "4            2.667474       0.0                 0.0   1934.344604  ...   \n",
       "2            3.530113       0.0                 0.0   2854.750488  ...   \n",
       "\n",
       "         ref  sc_value  side1_normalized  side1_score  side2_normalized  \\\n",
       "0  374.81054  0.491761         14.690318   837.348145         20.825792   \n",
       "3  374.09441  0.596517         14.805477   858.717651         20.484030   \n",
       "1  372.58994  0.468819         17.130129   907.896790         20.831068   \n",
       "4  374.83686  0.415013         17.054668   920.952087         22.980696   \n",
       "2  376.96998  0.464472         24.664927  1381.235962         33.255428   \n",
       "\n",
       "   side2_score  total_score  yhh_planarity  cdr_cluster_ID_L1  \\\n",
       "0   874.683289  2432.049006       0.134353            L1-11-1   \n",
       "3   860.329285  2397.836108       0.420196            L1-11-1   \n",
       "1   874.904846  2432.425954       0.061565            L1-11-1   \n",
       "4   965.189209  2600.787151       0.044588            L1-11-1   \n",
       "2  1396.727905  3441.859648       0.075642            L1-11-1   \n",
       "\n",
       "   cdr_cluster_ID_L3  \n",
       "0        L3-9-cis7-1  \n",
       "3        L3-9-cis7-1  \n",
       "1        L3-9-cis7-1  \n",
       "4        L3-9-cis7-1  \n",
       "2        L3-9-cis7-1  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = load_json_scorefile(\"expected_outputs/rabd/tutA1_score.sc\")\n",
    "df = drop_cluster_columns(df, keep_cdrs=[\"L1\", \"L3\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut A2. Graft Design\n",
    "\n",
    "Now we will be enabling graft design AND sequence design on L1 and L3 loops.  With an nstruct (n decoys) of 5, we are doing 25 design trials total - IE 25 actual grafts. \n",
    "\n",
    "\n",
    "    <AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" graft_design_cdrs=\"L1,L3\">\n",
    "\n",
    "\n",
    "This will take a about 2-3 times as long as sequence design, as grafting a non-breaking loop takes time.  This was 738 seconds on my laptop to generate 5. Here, you will generate 1 at about 150 seconds  Ouptut structures and scores are in ../expected_outputs/rabd.\n",
    "\n",
    "\t\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Typically, we require a much higher `-outer_cycle_rounds` and number of decoys to see anything significant.  Did this improve energies in your single antibody?  How about our pre-generated ones? Load and take a look at the scorefile as a pandas DataFrame as we did above (`expected_outputs/rabd/tutA2_score.sc`). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-6e0d61e14fe45905",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decoy</th>\n",
       "      <th>atom_pair_constraint</th>\n",
       "      <th>cdr_cluster_DIS_L1</th>\n",
       "      <th>cdr_cluster_DIS_L3</th>\n",
       "      <th>cdr_cluster_LEN_L1</th>\n",
       "      <th>cdr_cluster_LEN_L3</th>\n",
       "      <th>complex_normalized</th>\n",
       "      <th>dG_cross</th>\n",
       "      <th>dG_cross/dSASAx100</th>\n",
       "      <th>dG_separated</th>\n",
       "      <th>...</th>\n",
       "      <th>ref</th>\n",
       "      <th>sc_value</th>\n",
       "      <th>side1_normalized</th>\n",
       "      <th>side1_score</th>\n",
       "      <th>side2_normalized</th>\n",
       "      <th>side2_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>yhh_planarity</th>\n",
       "      <th>cdr_cluster_ID_L1</th>\n",
       "      <th>cdr_cluster_ID_L3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>tutA2_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.192533</td>\n",
       "      <td>31.677544</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.450019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1751.548462</td>\n",
       "      <td>...</td>\n",
       "      <td>364.65868</td>\n",
       "      <td>0.564960</td>\n",
       "      <td>15.000569</td>\n",
       "      <td>825.031311</td>\n",
       "      <td>21.473850</td>\n",
       "      <td>837.480164</td>\n",
       "      <td>2383.868655</td>\n",
       "      <td>0.107882</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>tutA2_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.217281</td>\n",
       "      <td>24.178181</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.760204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1843.231445</td>\n",
       "      <td>...</td>\n",
       "      <td>375.17336</td>\n",
       "      <td>0.537087</td>\n",
       "      <td>13.518681</td>\n",
       "      <td>851.676880</td>\n",
       "      <td>21.286449</td>\n",
       "      <td>979.176697</td>\n",
       "      <td>2707.759790</td>\n",
       "      <td>0.221997</td>\n",
       "      <td>L1-15-1</td>\n",
       "      <td>L3-11-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tutA2_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.599037</td>\n",
       "      <td>27.562218</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3.348572</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2484.706299</td>\n",
       "      <td>...</td>\n",
       "      <td>374.05882</td>\n",
       "      <td>0.486206</td>\n",
       "      <td>22.621260</td>\n",
       "      <td>1198.926758</td>\n",
       "      <td>29.544832</td>\n",
       "      <td>1270.427734</td>\n",
       "      <td>3268.206329</td>\n",
       "      <td>0.361406</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-10-cis7,8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>tutA2_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29.971132</td>\n",
       "      <td>45.571491</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.352178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2587.653564</td>\n",
       "      <td>...</td>\n",
       "      <td>373.00912</td>\n",
       "      <td>0.620726</td>\n",
       "      <td>18.605259</td>\n",
       "      <td>1265.157593</td>\n",
       "      <td>28.719591</td>\n",
       "      <td>1292.381592</td>\n",
       "      <td>3288.486156</td>\n",
       "      <td>0.126786</td>\n",
       "      <td>L1-17-1</td>\n",
       "      <td>L3-9-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>tutA2_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.153269</td>\n",
       "      <td>17.002886</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.869139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5082.230469</td>\n",
       "      <td>...</td>\n",
       "      <td>376.47676</td>\n",
       "      <td>0.571551</td>\n",
       "      <td>42.399551</td>\n",
       "      <td>2501.573486</td>\n",
       "      <td>62.364841</td>\n",
       "      <td>2494.593750</td>\n",
       "      <td>5710.672501</td>\n",
       "      <td>0.100791</td>\n",
       "      <td>L1-10-2</td>\n",
       "      <td>L3-8-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              decoy  atom_pair_constraint  cdr_cluster_DIS_L1  \\\n",
       "4  tutA2_my_ab_0005                   0.0           11.192533   \n",
       "2  tutA2_my_ab_0003                   0.0           20.217281   \n",
       "0  tutA2_my_ab_0001                   0.0            8.599037   \n",
       "1  tutA2_my_ab_0002                   0.0           29.971132   \n",
       "3  tutA2_my_ab_0004                   0.0           12.153269   \n",
       "\n",
       "   cdr_cluster_DIS_L3  cdr_cluster_LEN_L1  cdr_cluster_LEN_L3  \\\n",
       "4           31.677544                11.0                 7.0   \n",
       "2           24.178181                15.0                11.0   \n",
       "0           27.562218                11.0                10.0   \n",
       "1           45.571491                17.0                 9.0   \n",
       "3           17.002886                10.0                 8.0   \n",
       "\n",
       "   complex_normalized  dG_cross  dG_cross/dSASAx100  dG_separated  ...  \\\n",
       "4            2.450019       0.0                 0.0   1751.548462  ...   \n",
       "2            2.760204       0.0                 0.0   1843.231445  ...   \n",
       "0            3.348572       0.0                 0.0   2484.706299  ...   \n",
       "1            3.352178       0.0                 0.0   2587.653564  ...   \n",
       "3            5.869139       0.0                 0.0   5082.230469  ...   \n",
       "\n",
       "         ref  sc_value  side1_normalized  side1_score  side2_normalized  \\\n",
       "4  364.65868  0.564960         15.000569   825.031311         21.473850   \n",
       "2  375.17336  0.537087         13.518681   851.676880         21.286449   \n",
       "0  374.05882  0.486206         22.621260  1198.926758         29.544832   \n",
       "1  373.00912  0.620726         18.605259  1265.157593         28.719591   \n",
       "3  376.47676  0.571551         42.399551  2501.573486         62.364841   \n",
       "\n",
       "   side2_score  total_score  yhh_planarity  cdr_cluster_ID_L1  \\\n",
       "4   837.480164  2383.868655       0.107882            L1-11-1   \n",
       "2   979.176697  2707.759790       0.221997            L1-15-1   \n",
       "0  1270.427734  3268.206329       0.361406            L1-11-1   \n",
       "1  1292.381592  3288.486156       0.126786            L1-17-1   \n",
       "3  2494.593750  5710.672501       0.100791            L1-10-2   \n",
       "\n",
       "   cdr_cluster_ID_L3  \n",
       "4             L3-7-1  \n",
       "2            L3-11-1  \n",
       "0     L3-10-cis7,8-1  \n",
       "1             L3-9-1  \n",
       "3             L3-8-1  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### BEGIN SOLUTION\n",
    "\n",
    "df_a2 = load_json_scorefile(\"expected_outputs/rabd/tutA2_score.sc\")\n",
    "df_a2 = drop_cluster_columns(df_a2, keep_cdrs=[\"L1\", \"L3\"])\n",
    "df_a2\n",
    "\n",
    "### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets merge these dataframes, sort by dG_separated, and see if any of our graft-design models did better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decoy</th>\n",
       "      <th>atom_pair_constraint</th>\n",
       "      <th>cdr_cluster_DIS_L1</th>\n",
       "      <th>cdr_cluster_DIS_L3</th>\n",
       "      <th>cdr_cluster_LEN_L1</th>\n",
       "      <th>cdr_cluster_LEN_L3</th>\n",
       "      <th>complex_normalized</th>\n",
       "      <th>dG_cross</th>\n",
       "      <th>dG_cross/dSASAx100</th>\n",
       "      <th>dG_separated</th>\n",
       "      <th>...</th>\n",
       "      <th>ref</th>\n",
       "      <th>sc_value</th>\n",
       "      <th>side1_normalized</th>\n",
       "      <th>side1_score</th>\n",
       "      <th>side2_normalized</th>\n",
       "      <th>side2_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>yhh_planarity</th>\n",
       "      <th>cdr_cluster_ID_L1</th>\n",
       "      <th>cdr_cluster_ID_L3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>tutA2_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.192533</td>\n",
       "      <td>31.677544</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.450019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1751.548462</td>\n",
       "      <td>...</td>\n",
       "      <td>364.65868</td>\n",
       "      <td>0.564960</td>\n",
       "      <td>15.000569</td>\n",
       "      <td>825.031311</td>\n",
       "      <td>21.473850</td>\n",
       "      <td>837.480164</td>\n",
       "      <td>2383.868655</td>\n",
       "      <td>0.107882</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tutA1_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.244154</td>\n",
       "      <td>14.169669</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1784.542480</td>\n",
       "      <td>...</td>\n",
       "      <td>374.81054</td>\n",
       "      <td>0.491761</td>\n",
       "      <td>14.690318</td>\n",
       "      <td>837.348145</td>\n",
       "      <td>20.825792</td>\n",
       "      <td>874.683289</td>\n",
       "      <td>2432.049006</td>\n",
       "      <td>0.134353</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>tutA1_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.604385</td>\n",
       "      <td>10.078285</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.459319</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1820.213013</td>\n",
       "      <td>...</td>\n",
       "      <td>374.09441</td>\n",
       "      <td>0.596517</td>\n",
       "      <td>14.805477</td>\n",
       "      <td>858.717651</td>\n",
       "      <td>20.484030</td>\n",
       "      <td>860.329285</td>\n",
       "      <td>2397.836108</td>\n",
       "      <td>0.420196</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>tutA2_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.217281</td>\n",
       "      <td>24.178181</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.760204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1843.231445</td>\n",
       "      <td>...</td>\n",
       "      <td>375.17336</td>\n",
       "      <td>0.537087</td>\n",
       "      <td>13.518681</td>\n",
       "      <td>851.676880</td>\n",
       "      <td>21.286449</td>\n",
       "      <td>979.176697</td>\n",
       "      <td>2707.759790</td>\n",
       "      <td>0.221997</td>\n",
       "      <td>L1-15-1</td>\n",
       "      <td>L3-11-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>tutA1_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.939879</td>\n",
       "      <td>14.834209</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.494796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1849.303833</td>\n",
       "      <td>...</td>\n",
       "      <td>372.58994</td>\n",
       "      <td>0.468819</td>\n",
       "      <td>17.130129</td>\n",
       "      <td>907.896790</td>\n",
       "      <td>20.831068</td>\n",
       "      <td>874.904846</td>\n",
       "      <td>2432.425954</td>\n",
       "      <td>0.061565</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>tutA1_my_ab_0005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.124744</td>\n",
       "      <td>14.362237</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.667474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1934.344604</td>\n",
       "      <td>...</td>\n",
       "      <td>374.83686</td>\n",
       "      <td>0.415013</td>\n",
       "      <td>17.054668</td>\n",
       "      <td>920.952087</td>\n",
       "      <td>22.980696</td>\n",
       "      <td>965.189209</td>\n",
       "      <td>2600.787151</td>\n",
       "      <td>0.044588</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>tutA2_my_ab_0001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.599037</td>\n",
       "      <td>27.562218</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3.348572</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2484.706299</td>\n",
       "      <td>...</td>\n",
       "      <td>374.05882</td>\n",
       "      <td>0.486206</td>\n",
       "      <td>22.621260</td>\n",
       "      <td>1198.926758</td>\n",
       "      <td>29.544832</td>\n",
       "      <td>1270.427734</td>\n",
       "      <td>3268.206329</td>\n",
       "      <td>0.361406</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-10-cis7,8-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>tutA2_my_ab_0002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29.971132</td>\n",
       "      <td>45.571491</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.352178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2587.653564</td>\n",
       "      <td>...</td>\n",
       "      <td>373.00912</td>\n",
       "      <td>0.620726</td>\n",
       "      <td>18.605259</td>\n",
       "      <td>1265.157593</td>\n",
       "      <td>28.719591</td>\n",
       "      <td>1292.381592</td>\n",
       "      <td>3288.486156</td>\n",
       "      <td>0.126786</td>\n",
       "      <td>L1-17-1</td>\n",
       "      <td>L3-9-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>tutA1_my_ab_0003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.693962</td>\n",
       "      <td>12.150679</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.530113</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2854.750488</td>\n",
       "      <td>...</td>\n",
       "      <td>376.96998</td>\n",
       "      <td>0.464472</td>\n",
       "      <td>24.664927</td>\n",
       "      <td>1381.235962</td>\n",
       "      <td>33.255428</td>\n",
       "      <td>1396.727905</td>\n",
       "      <td>3441.859648</td>\n",
       "      <td>0.075642</td>\n",
       "      <td>L1-11-1</td>\n",
       "      <td>L3-9-cis7-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>tutA2_my_ab_0004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.153269</td>\n",
       "      <td>17.002886</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.869139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5082.230469</td>\n",
       "      <td>...</td>\n",
       "      <td>376.47676</td>\n",
       "      <td>0.571551</td>\n",
       "      <td>42.399551</td>\n",
       "      <td>2501.573486</td>\n",
       "      <td>62.364841</td>\n",
       "      <td>2494.593750</td>\n",
       "      <td>5710.672501</td>\n",
       "      <td>0.100791</td>\n",
       "      <td>L1-10-2</td>\n",
       "      <td>L3-8-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              decoy  atom_pair_constraint  cdr_cluster_DIS_L1  \\\n",
       "5  tutA2_my_ab_0005                   0.0           11.192533   \n",
       "0  tutA1_my_ab_0001                   0.0           10.244154   \n",
       "1  tutA1_my_ab_0004                   0.0            9.604385   \n",
       "6  tutA2_my_ab_0003                   0.0           20.217281   \n",
       "2  tutA1_my_ab_0002                   0.0            9.939879   \n",
       "3  tutA1_my_ab_0005                   0.0           13.124744   \n",
       "7  tutA2_my_ab_0001                   0.0            8.599037   \n",
       "8  tutA2_my_ab_0002                   0.0           29.971132   \n",
       "4  tutA1_my_ab_0003                   0.0            9.693962   \n",
       "9  tutA2_my_ab_0004                   0.0           12.153269   \n",
       "\n",
       "   cdr_cluster_DIS_L3  cdr_cluster_LEN_L1  cdr_cluster_LEN_L3  \\\n",
       "5           31.677544                11.0                 7.0   \n",
       "0           14.169669                11.0                 9.0   \n",
       "1           10.078285                11.0                 9.0   \n",
       "6           24.178181                15.0                11.0   \n",
       "2           14.834209                11.0                 9.0   \n",
       "3           14.362237                11.0                 9.0   \n",
       "7           27.562218                11.0                10.0   \n",
       "8           45.571491                17.0                 9.0   \n",
       "4           12.150679                11.0                 9.0   \n",
       "9           17.002886                10.0                 8.0   \n",
       "\n",
       "   complex_normalized  dG_cross  dG_cross/dSASAx100  dG_separated  ...  \\\n",
       "5            2.450019       0.0                 0.0   1751.548462  ...   \n",
       "0            2.494409       0.0                 0.0   1784.542480  ...   \n",
       "1            2.459319       0.0                 0.0   1820.213013  ...   \n",
       "6            2.760204       0.0                 0.0   1843.231445  ...   \n",
       "2            2.494796       0.0                 0.0   1849.303833  ...   \n",
       "3            2.667474       0.0                 0.0   1934.344604  ...   \n",
       "7            3.348572       0.0                 0.0   2484.706299  ...   \n",
       "8            3.352178       0.0                 0.0   2587.653564  ...   \n",
       "4            3.530113       0.0                 0.0   2854.750488  ...   \n",
       "9            5.869139       0.0                 0.0   5082.230469  ...   \n",
       "\n",
       "         ref  sc_value  side1_normalized  side1_score  side2_normalized  \\\n",
       "5  364.65868  0.564960         15.000569   825.031311         21.473850   \n",
       "0  374.81054  0.491761         14.690318   837.348145         20.825792   \n",
       "1  374.09441  0.596517         14.805477   858.717651         20.484030   \n",
       "6  375.17336  0.537087         13.518681   851.676880         21.286449   \n",
       "2  372.58994  0.468819         17.130129   907.896790         20.831068   \n",
       "3  374.83686  0.415013         17.054668   920.952087         22.980696   \n",
       "7  374.05882  0.486206         22.621260  1198.926758         29.544832   \n",
       "8  373.00912  0.620726         18.605259  1265.157593         28.719591   \n",
       "4  376.96998  0.464472         24.664927  1381.235962         33.255428   \n",
       "9  376.47676  0.571551         42.399551  2501.573486         62.364841   \n",
       "\n",
       "   side2_score  total_score  yhh_planarity  cdr_cluster_ID_L1  \\\n",
       "5   837.480164  2383.868655       0.107882            L1-11-1   \n",
       "0   874.683289  2432.049006       0.134353            L1-11-1   \n",
       "1   860.329285  2397.836108       0.420196            L1-11-1   \n",
       "6   979.176697  2707.759790       0.221997            L1-15-1   \n",
       "2   874.904846  2432.425954       0.061565            L1-11-1   \n",
       "3   965.189209  2600.787151       0.044588            L1-11-1   \n",
       "7  1270.427734  3268.206329       0.361406            L1-11-1   \n",
       "8  1292.381592  3288.486156       0.126786            L1-17-1   \n",
       "4  1396.727905  3441.859648       0.075642            L1-11-1   \n",
       "9  2494.593750  5710.672501       0.100791            L1-10-2   \n",
       "\n",
       "   cdr_cluster_ID_L3  \n",
       "5             L3-7-1  \n",
       "0        L3-9-cis7-1  \n",
       "1        L3-9-cis7-1  \n",
       "6            L3-11-1  \n",
       "2        L3-9-cis7-1  \n",
       "3        L3-9-cis7-1  \n",
       "7     L3-10-cis7,8-1  \n",
       "8             L3-9-1  \n",
       "4        L3-9-cis7-1  \n",
       "9             L3-8-1  \n",
       "\n",
       "[10 rows x 48 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tut_a12 = pandas.concat([df, df_a2], ignore_index=True).sort_values(\"dG_separated\", ascending=True)\n",
    "df_tut_a12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take a look at the lowest (dG) scoring pose in pymol - do you see any difference in L1 and L3 loops there?  Do they make better contact than what we had before?\n",
    "\n",
    "\tLets take a look in pymol.   \n",
    "\n",
    "\t\t\tpymol inputs/rabd/my_ab.pdb inputs/rabd/tutA2_* \n",
    "\t\t\t@color_cdrs.pml\n",
    "\t\t\tcenter full_epitope\n",
    "\n",
    " How different are the L1 and L3 loops? Have any changed length?\n",
    "\n",
    " Lets take a look at the clusters in our dataframe.  Have they changed from the native?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    print(\"L1\", original_pose.scores[\"cdr_cluster_ID_L1\"])\n",
    "    print(\"L3\", original_pose.scores[\"cdr_cluster_ID_L3\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut A3.  Basic De-novo run\n",
    "\n",
    "Here, we want to do a denovo-run (without docking), starting with random CDRs grafted in - instead of whatever we have in the antibody to start with (only for the CDRs that are actually undergoing graft-design).  This is useful, as we start the design with very high energy and work our way down.  Note that since this is an entirely new interface for our model protein, this interface is already at a very high energy - and so this is less needed, but it should be noted how to do this. (139 seconds on my laptop). Do this below as you have done in other tutorials - either through code or XML.\n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" graft_design_cdrs=\"L1,L3\" seq_design_cdrs=\"L1,L3\" \n",
    "\t\t\t                                                                  random_start=\"1\"/>  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-b7dcc3797eed573e",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    ### BEGIN SOLUTION\n",
    "    pose = original_pose.clone()\n",
    "    rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" graft_design_cdrs=\"L1,L3\" seq_design_cdrs=\"L1,L3\" random_start=\"1\"/> light_chain=\"kappa\"')\n",
    "    rabd.apply(pose)\n",
    "\n",
    "    # OR (REUSE code from above)\n",
    "    pose = original_pose.clone()\n",
    "    rabd2.set_seq_design_cdrs(cdrs)\n",
    "    rabd2.set_graft_design_cdrs(cdrs)\n",
    "    rabd2.set_random_start(True)\n",
    "    rabd2.set_light_chain(\"kappa\")\n",
    "    rabd2.apply(pose)\n",
    "\n",
    "    ### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would starting from a random CDR help anywhere?  Perhaps if you want an entirely new cluster or length to break a patent or remove some off target effects?  We will use it below to start de novo design with docking."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut A4. RAbD Framework Components\n",
    "\n",
    "This tutorial will give you some exprience with an antibody design protocol using the `RosettaAntibdyDesign` components.  We will take the light chain CDRs from a malaria antibody and graft them into our antibody.  In the tutorial we are interested in stabilizing the grafted CDRs in relation to the whole antibody, instead of interface design to an antigen.  \n",
    "\n",
    "We will graft the CDRs in, minimize the structure with CDR dihedral constraints (that use the CDR clusters) to not purturb the CDRs too much, and then design the framework around the CDRs while designing the CDRs and neighbors.  The result should be mutations that better accomodate our new CDRs.  This can be useful for humanizing potential antibodies or framework switching, where we want the binding properties of certain CDRs, but the stability or immunological profile of a different framework.  \n",
    "\n",
    "We are using an XML here for simplicity - all components are available in PyRosetta, but harder to setup. \n",
    "\n",
    "#### 1. Copy the Files\n",
    "\n",
    "        cp ../inputs/rabd/ab_design_components.xml .\n",
    "\t\tcp ../inputs/rabd/malaria_cdrs.pdb .\n",
    "\t\n",
    "\t\t\n",
    "Take a look at the xml.  \n",
    "\n",
    "- We are using the `AntibodyCDRGrafter` to do the grafting of our CDRs.  \n",
    "- We then add constraints using `CDRDihderalConstraintMovers` for each CDR with use the CDR clusters determinedy by RosettaAntibody to keep from perturbing the CDRs too much.  \n",
    "- Finally, we do a round of pack/min/pack using the `RestrictToCDRsAndNeighborsOperation` and the `CDRResidueSelector`.  This task operation controls what we pack and design.  It first limits packing and design to only the CDRs and its neighbors.  By specifying the `design_framework=1` option we allow the neighbor framework residues to design, while the CDRs and antigen neighbors will only repack.  If we wanted to disable antigen repacking, we would pass the _DisableAntibodyRegionOperation_ task operation.  Using this, we can specify any antibody region as `antibody_region`, `cdr_region`, or `antigen_region` and we can disable just design or both packing and design.  \n",
    "\t\t\n",
    "\n",
    "These task operations allow us to chisel exactly what we want to design in antibody, sans a residue-specific resfile (though we could combine these with one of them!).  All of these tools are available in-code.  If you've done the design workshop, you will know how to use them here. Checkout `rosetta.protocols.antibody.task_operations` for a list of them.\n",
    "\t\t\n",
    "Finally, we use the new SimpleMetric system to obtain our final sequence of the CDRs to compare to our native antibody as well as pymol selections of our CDRs - which you have been introduced to in the previous tutorial. \n",
    "\n",
    "\n",
    "##### PyRosetta Locations\n",
    "\n",
    "`rosetta.protocols.antibody.task_operations`\n",
    "\n",
    "`rosetta.protocols.antibody.constraints`\n",
    "\n",
    "`rosetta.protocols.antibody.residue_selectors`\n",
    "\n",
    "##### Documentation\n",
    "\n",
    "- <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/SimpleMetrics/SimpleMetrics>\n",
    "\n",
    "- More Documentation is available here:\n",
    "\n",
    " - <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/RosettaScripts>\n",
    "\n",
    " - <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/TaskOperations/TaskOperations-RosettaScripts#antibody-and-cdr-specific-operations>\n",
    "\n",
    " - <https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/Movers-RosettaScripts#antibody-modeling-and-design-movers>\n",
    "\t\n",
    "#### 2. Run the protocol or copy the output (357 seconds).  \n",
    "\n",
    "\n",
    "#### 3. Look at the score file as you have before.  Are the sequences different between what we started with?  How about the interaction energies?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system('cp inputs/rabd/malaria_cdrs.pdb .')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mGenerating XML Schema for rosetta_scripts...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mInitializing schema validator...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mValidating input script...\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0m...done\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mParsed script:\n",
      "<ROSETTASCRIPTS>\n",
      "\t<RESIDUE_SELECTORS>\n",
      "\t\t<CDR cdrs=\"L1,L2,L3\" name=\"light_cdrs\"/>\n",
      "\t\t<CDR cdrs=\"L1\" name=\"L1\"/>\n",
      "\t\t<CDR cdrs=\"L2\" name=\"L2\"/>\n",
      "\t\t<CDR cdrs=\"L3\" name=\"L3\"/>\n",
      "\t\t<AntibodyRegion name=\"antigen\" region=\"antigen_region\"/>\n",
      "\t\t<AntibodyRegion name=\"framework\" region=\"framework_region\"/>\n",
      "\t\t<AntibodyRegion name=\"cdrs\" region=\"cdr_region\"/>\n",
      "\t</RESIDUE_SELECTORS>\n",
      "\t<TASKOPERATIONS>\n",
      "\t\t<RestrictToCDRsAndNeighbors cdrs=\"L1,L2,L3\" design_cdrs=\"1\" design_framework=\"1\" name=\"restrict_to_cdrs\"/>\n",
      "\t\t<AddCDRProfilesOperation cdrs=\"L1,L2,L3\" fallback_strategy=\"CONSERVATIVE\" name=\"profiles\"/>\n",
      "\t</TASKOPERATIONS>\n",
      "\t<MOVE_MAP_FACTORIES>\n",
      "\t\t<MoveMapFactory bb=\"0\" chi=\"0\" name=\"movemap_cdrs\">\n",
      "\t\t\t<Backbone residue_selector=\"light_cdrs\"/>\n",
      "\t\t\t<Chi residue_selector=\"light_cdrs\"/>\n",
      "\t\t</MoveMapFactory>\n",
      "\t</MOVE_MAP_FACTORIES>\n",
      "\t<SIMPLE_METRICS>\n",
      "\t\t<SasaMetric name=\"sasa\" residue_selector=\"light_cdrs\"/>\n",
      "\t\t<SelectedResiduesPyMOLMetric name=\"cdr_selection\" residue_selector=\"light_cdrs\"/>\n",
      "\t\t<SequenceMetric custom_type=\"L1\" name=\"L1_seq\" residue_selector=\"L1\"/>\n",
      "\t\t<SequenceMetric custom_type=\"L2\" name=\"L2_seq\" residue_selector=\"L2\"/>\n",
      "\t\t<SequenceMetric custom_type=\"L3\" name=\"L3_seq\" residue_selector=\"L3\"/>\n",
      "\t\t<InteractionEnergyMetric name=\"cdr-int\" residue_selector=\"light_cdrs\" residue_selector2=\"antigen\"/>\n",
      "\t</SIMPLE_METRICS>\n",
      "\t<MOVERS>\n",
      "\t\t<CDRDihedralConstraintMover cdr=\"L1\" name=\"dih_mover_L1\" use_cluster_csts=\"1\"/>\n",
      "\t\t<CDRDihedralConstraintMover cdr=\"L2\" name=\"dih_mover_L2\" use_cluster_csts=\"1\"/>\n",
      "\t\t<CDRDihedralConstraintMover cdr=\"L3\" name=\"dih_mover_L3\" use_cluster_csts=\"1\"/>\n",
      "\t\t<AntibodyCDRGrafter cdr_definition=\"North\" cdrs=\"L1,L2,L3\" donor_structure_from_pdb=\"malaria_cdrs.pdb\" input_ab_scheme=\"AHO_Scheme\" name=\"grafter\" optimize_cdr4_if_neighbor=\"1\" optimize_cdrs=\"0\" use_secondary_graft_mover=\"1\"/>\n",
      "\t\t<PackRotamersMover name=\"packrot\" task_operations=\"restrict_to_cdrs,profiles\"/>\n",
      "\t\t<MinMover cartesian=\"1\" movemap_factory=\"movemap_cdrs\" name=\"minmover\" tolerance=\".1\"/>\n",
      "\t\t<RunSimpleMetrics metrics=\"sasa,cdr_selection,L1_seq,L2_seq,L3_seq,cdr-int\" name=\"design_metrics\" prefix=\"design_\"/>\n",
      "\t\t<RunSimpleMetrics metrics=\"sasa,cdr_selection,L1_seq,L2_seq,L3_seq,cdr-int\" name=\"native_metrics\" prefix=\"native_\"/>\n",
      "\t</MOVERS>\n",
      "\t<PROTOCOLS>\n",
      "\t\t<Add mover_name=\"native_metrics\"/>\n",
      "\t\t<Add mover_name=\"grafter\"/>\n",
      "\t\t<Add mover_name=\"dih_mover_L1\"/>\n",
      "\t\t<Add mover_name=\"dih_mover_L2\"/>\n",
      "\t\t<Add mover_name=\"dih_mover_L3\"/>\n",
      "\t\t<Add mover_name=\"packrot\"/>\n",
      "\t\t<Add mover_name=\"minmover\"/>\n",
      "\t\t<Add mover_name=\"packrot\"/>\n",
      "\t\t<Add mover_name=\"minmover\"/>\n",
      "\t\t<Add mover_name=\"design_metrics\"/>\n",
      "\t</PROTOCOLS>\n",
      "</ROSETTASCRIPTS>\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.antibody.residue_selector.CDRResidueSelector: \u001b[0mSetting CDRs from settings\n",
      "\u001b[0mprotocols.antibody.residue_selector.CDRResidueSelector: \u001b[0mSetting CDRs from settings\n",
      "\u001b[0mprotocols.antibody.residue_selector.CDRResidueSelector: \u001b[0mSetting CDRs from settings\n",
      "\u001b[0mprotocols.antibody.residue_selector.CDRResidueSelector: \u001b[0mSetting CDRs from settings\n",
      "\u001b[0mprotocols.jd2.parser.TaskOperationLoader: \u001b[0mDefined TaskOperation named \"restrict_to_cdrs\" of type RestrictToCDRsAndNeighbors\n",
      "\u001b[0mprotocols.task_operations.ConservativeDesignOperation: \u001b[0mLoading conservative mutational data\n",
      "\u001b[0mprotocols.antibody.task_operations.AddCDRProfilesOperation: \u001b[0mSetting CDRs from settings\n",
      "\u001b[0mprotocols.jd2.parser.TaskOperationLoader: \u001b[0mDefined TaskOperation named \"profiles\" of type AddCDRProfilesOperation\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector light_cdrs\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector light_cdrs\n",
      "\u001b[0mprotocols.jd2.parser.MoveMapFactoryLoader: \u001b[0mDefined MoveMap named \"movemap_cdrs\"\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector light_cdrs\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector light_cdrs\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector L1\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector L2\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector L3\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector light_cdrs\n",
      "\u001b[0mcore.select.residue_selector.util: \u001b[0mFound residue selector antigen\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"dih_mover_L1\" of type CDRDihedralConstraintMover\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"dih_mover_L2\" of type CDRDihedralConstraintMover\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"dih_mover_L3\" of type CDRDihedralConstraintMover\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mSetting CDRs from settings\n",
      "\u001b[0mcore.import_pose.import_pose: \u001b[0mFile 'malaria_cdrs.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.io.pdb.pdb_reader: \u001b[0mParsing 0 .pdb records with unknown format to search for Rosetta-specific comments.\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m missing heavyatom:  OXT on residue CYS:CtermProteinFull 387\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m missing heavyatom:  OXT on residue ASP:CtermProteinFull 601\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 8 22\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 8 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 22 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 8 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 22 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 24 36\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 24 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 36 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 24 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 36 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 43 58\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 43 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 58 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 43 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 58 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 52 70\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 52 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 70 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 52 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 70 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 72 83\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 72 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 72 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 88 98\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 88 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 98 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 88 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 98 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 93 111\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 93 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 111 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 93 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 111 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 113 127\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 113 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 127 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 113 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 127 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 135 146\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 135 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 146 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 135 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 146 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 139 155\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 139 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 155 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 139 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 155 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 157 170\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 157 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 170 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 157 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 170 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 197 261\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 197 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 261 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 197 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 261 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 307 367\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 307 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 367 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 307 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 367 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 407 481\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 407 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 481 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 407 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 481 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 527 582\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 527 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 582 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 527 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 582 CYD\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"grafter\" of type AntibodyCDRGrafter\n",
      "\u001b[0mcore.pack.task.xml_util: \u001b[0mObject packrot reading the following task_operations: Adding the following task operations\n",
      "restrict_to_cdrs profiles\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"packrot\" of type PackRotamersMover\n",
      "\u001b[0mprotocols.minimization_packing.MinMover: \u001b[0mFound set MoveMap factory. Using this to define MoveMap.\n",
      "\u001b[0mcore.select.movemap.util: \u001b[0mFound MoveMapFactory movemap_cdrs\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"minmover\" of type MinMover\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SasaMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SelectedResiduesPyMOLMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SequenceMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SequenceMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SequenceMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric InteractionEnergyMetric.\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"design_metrics\" of type RunSimpleMetrics\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SasaMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SelectedResiduesPyMOLMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SequenceMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SequenceMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric SequenceMetric.\n",
      "\u001b[0mcore.simple_metrics.util: \u001b[0mAdded simple metric InteractionEnergyMetric.\n",
      "\u001b[0mprotocols.rosetta_scripts.RosettaScriptsParser: \u001b[0mDefined mover named \"native_metrics\" of type RunSimpleMetrics\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0mParsedProtocol mover with the following movers and filters\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"native_metrics\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"grafter\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"dih_mover_L1\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"dih_mover_L2\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"dih_mover_L3\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"packrot\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"minmover\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"packrot\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"minmover\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0madded mover \"design_metrics\" with filter \"true_filter\"\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER RunSimpleMetrics - native_metrics=======================\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SasaMetric - calculating sasa\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SelectedResiduesPyMOLMetric - calculating pymol_selection\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L1_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L2_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L3_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: InteractionEnergyMetric - calculating interaction_energy\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m ################ Cloning pose and Scoring! ##############################\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Ensure that pose is scored\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m before using InteractionEnergyMetric for maximum performance!\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m ##########################################################################\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_atr 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -74.0845\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -74.0845\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_rep 0.55\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 11946.9\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 6570.79\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_sol 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 58.1828\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 58.1828\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mlk_ball_wtd 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 3.26267\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 3.26267\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_elec 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -12.7929\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -12.7929\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mhbond_bb_sc 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -0.53415\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -0.53415\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER AntibodyCDRGrafter - grafter=======================\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGrafting CDR: L1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mReturning 16 residues from 195 to 210\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mReverting out-of-date disulfide to thiol type at resid 3\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 891 End: 903 NterO: 3 CterO: 3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mSuperimposing overhang residues\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 11 residues from 892 to 902\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 945\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 945 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 945 CYD\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 10\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 891 -1  EDGE 891 892 1  EDGE 892 964 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 891\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mOriginal End: 903\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mEnd: 902\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mInsert Length: 10\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/centroid_smooth/cen_smooth_params.txt\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 890  stop: 893  cut: 891  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 900  stop: 903  cut: 901  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.loops.FoldTreeFromLoopsWrapper: \u001b[0mold foldtree FOLD_TREE  EDGE 1 891 -1  EDGE 891 902 1  EDGE 902 974 -1  EDGE 891 892 -2  C    N    EDGE 892 893 -2  C    N    EDGE 893 894 -2  C    N    EDGE 894 895 -2  C    N    EDGE 895 896 -2  C    N    EDGE 896 897 -2  C    N    EDGE 897 898 -2  C    N    EDGE 898 899 -2  C    N    EDGE 899 900 -2  C    N    EDGE 900 901 -2  C    N   \n",
      "New foldtree FOLD_TREE  EDGE 1 889 -1  EDGE 889 891 -1  EDGE 889 894 1  EDGE 894 892 -1  EDGE 894 899 -1  EDGE 899 901 -1  EDGE 899 904 2  EDGE 904 902 -1  EDGE 904 974 -1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  890\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 893\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 900\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 902\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 903\n",
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mFinished initializing centroid residue type set.  Created 62 residue types\n",
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mTotal time to initialize 0.027801 seconds.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_distance_score\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_CaCbCb_angle_score\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_CaCbCbCa_dihedral_score\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_backbone_dihedral_score\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mstart 1661.69\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 1\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60113 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m1 1344.14\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 2\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62071 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 3\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61218 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m3 1338.91\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 4\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.68362 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m4 1320.84\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 5\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.76509 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 6\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.65825 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m6 1313.39\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 7\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60819 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 8\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62875 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 9\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6266 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 10\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60916 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 11\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61684 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 12\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61848 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 13\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61645 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 14\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61543 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 15\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62401 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 16\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.63562 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 17\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.64733 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 18\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60473 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 19\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60487 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 20\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6193 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 21\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60541 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 22\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60568 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 23\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62672 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 24\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.5793 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 25\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.5867 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 26\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59485 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 27\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.64499 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 28\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61149 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 29\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.63725 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 30\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61434 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 31\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6164 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 32\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6038 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 33\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61122 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 34\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59831 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 35\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59755 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 36\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60123 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 37\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59705 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 38\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61627 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 39\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61896 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 40\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60807 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 41\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60502 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 42\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62934 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 43\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62077 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 44\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61599 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 45\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6229 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 46\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61137 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 47\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.57139 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 48\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.5991 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 49\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6038 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 50\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60968 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 51\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.58477 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 52\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.56761 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 53\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.63019 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 54\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59919 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 55\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62732 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 56\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.64698 max: 1.5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 57\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61982 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m57 1312.52\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 58\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60883 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 59\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59947 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 60\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.63073 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 61\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60254 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 62\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61244 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 63\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61427 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 64\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62712 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 65\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61469 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 66\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60174 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 67\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61892 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 68\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62348 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 69\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62007 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 70\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62511 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 71\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.5895 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 72\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60284 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 73\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59508 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 74\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.58678 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 75\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.63562 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 76\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59452 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 77\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.62141 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 78\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6158 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 79\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60115 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 80\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61093 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 81\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61939 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 82\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60062 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 83\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60587 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 84\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59995 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 85\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61391 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 86\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.60314 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 87\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61951 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 88\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61404 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 89\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61096 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 90\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59714 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 91\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61394 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 92\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61956 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 93\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59923 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 94\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61766 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 95\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.59173 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 96\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6177 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 97\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.6377 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 98\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61976 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 99\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61219 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 100\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.58123 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mfinish 1312.52\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 151 rotamers at 19 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61982 max: 1.5\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft meets ideal geometry false\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 42 L  C-N --  1.61982 max: 1.5\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGraft not fully closed. Using secondary graft mover\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 11 residues from 892 to 902\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 945\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 945 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 945 CYD\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 10\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 891 -1  EDGE 891 892 1  EDGE 892 964 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 889 -1  EDGE 889 902 -1  EDGE 889 904 1  EDGE 904 903 -1  EDGE 904 974 -1\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mLoops: LOOP  begin  end  cut  skip_rate  extended\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mLOOP start: 890  stop: 903  cut: 902  size: 14  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 902\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  890\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 893\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 900\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 902\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 903\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mstart 3382.37\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mround 1\n",
      "\u001b[0mcore.optimization.LineMinimizer: \u001b[0m\u001b[31m\u001b[1m[ ERROR ]\u001b[0m \u001b[31mInaccurate G! step= 1.90735e-06 Deriv= -12.3575 Finite Diff= 4.77606\u001b[0m\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 902 - 43 L  Ca-C-N --  99.5 (min): 139.257 : 134.5 (max)\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0m1 1395.71\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mround 2\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mGraft Closed early - returning\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0m2 1299.88\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mfinish 1299.88\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 151 rotamers at 19 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mGraft meets ideal geometry true\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGrafting CDR: L2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mReturning 14 residues from 219 to 232\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 915 End: 924 NterO: 3 CterO: 3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mSuperimposing overhang residues\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 8 residues from 916 to 923\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mReverting out-of-date disulfide to thiol type at resid 891\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mReverting out-of-date disulfide to thiol type at resid 947\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 915\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 916\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 8\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 915 -1  EDGE 915 916 1  EDGE 916 966 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 915\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mOriginal End: 924\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mEnd: 924\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mInsert Length: 8\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 914  stop: 917  cut: 915  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 922  stop: 925  cut: 923  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.loops.FoldTreeFromLoopsWrapper: \u001b[0mold foldtree FOLD_TREE  EDGE 1 915 -1  EDGE 915 924 1  EDGE 924 974 -1  EDGE 915 916 -2  C    N    EDGE 916 917 -2  C    N    EDGE 917 918 -2  C    N    EDGE 918 919 -2  C    N    EDGE 919 920 -2  C    N    EDGE 920 921 -2  C    N    EDGE 921 922 -2  C    N    EDGE 922 923 -2  C    N   \n",
      "New foldtree FOLD_TREE  EDGE 1 913 -1  EDGE 913 915 -1  EDGE 913 918 1  EDGE 918 916 -1  EDGE 918 921 -1  EDGE 921 923 -1  EDGE 921 926 2  EDGE 926 924 -1  EDGE 926 974 -1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 915\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  914\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  915\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 916\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 916\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 917\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 922\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 924\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 925\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mstart 1712.66\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 1\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 923 - 72 L  Ca-C-N --  99.5 (min): 147.941 : 134.5 (max)\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m1 1351.45\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 2\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft Closed early - returning\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m2 1256.39\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mfinish 1256.39\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 148 rotamers at 14 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft meets ideal geometry true\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mSuccess.  Graft closed.\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGrafting CDR: L3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mReturning 15 residues from 259 to 273\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mReverting out-of-date disulfide to thiol type at resid 3\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 955 End: 965 NterO: 3 CterO: 3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mSuperimposing overhang residues\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 9 residues from 956 to 964\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 955\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 956\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 9\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 955 -1  EDGE 955 956 1  EDGE 956 965 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 955\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mOriginal End: 965\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mEnd: 965\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mInsert Length: 9\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 954  stop: 957  cut: 955  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 963  stop: 966  cut: 964  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.loops.FoldTreeFromLoopsWrapper: \u001b[0mold foldtree FOLD_TREE  EDGE 1 955 -1  EDGE 955 965 1  EDGE 965 974 -1  EDGE 955 956 -2  C    N    EDGE 956 957 -2  C    N    EDGE 957 958 -2  C    N    EDGE 958 959 -2  C    N    EDGE 959 960 -2  C    N    EDGE 960 961 -2  C    N    EDGE 961 962 -2  C    N    EDGE 962 963 -2  C    N    EDGE 963 964 -2  C    N   \n",
      "New foldtree FOLD_TREE  EDGE 1 953 -1  EDGE 953 955 -1  EDGE 953 958 1  EDGE 958 956 -1  EDGE 958 962 -1  EDGE 962 964 -1  EDGE 962 967 2  EDGE 967 965 -1  EDGE 967 974 -1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 955\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  954\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  955\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 956\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 956\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 957\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 963\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 965\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 966\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mstart 1638.41\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 1\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft Closed early - returning\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m1 1265.74\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mfinish 1265.74\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 146 rotamers at 19 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft meets ideal geometry true\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mSuccess.  Graft closed.\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER CDRDihedralConstraintMover - dih_mover_L1=======================\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.constraints.CDRDihedralConstraintMover: \u001b[0mL1-10-1 data: 42 cutoff: 10\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mread constraints from /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/sampling/antibodies/cluster_based_constraints/CircularHarmonic/outliers_false_liberal_use_means/L1-10-1.txt\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mRead in 20 constraints\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER CDRDihedralConstraintMover - dih_mover_L2=======================\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.constraints.CDRDihedralConstraintMover: \u001b[0mL2-8-1 data: 632 cutoff: 10\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mread constraints from /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/sampling/antibodies/cluster_based_constraints/CircularHarmonic/outliers_false_liberal_use_means/L2-8-1.txt\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mRead in 16 constraints\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER CDRDihedralConstraintMover - dih_mover_L3=======================\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.constraints.CDRDihedralConstraintMover: \u001b[0mL3-9-cis7-1 data: 419 cutoff: 10\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mread constraints from /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database/sampling/antibodies/cluster_based_constraints/CircularHarmonic/outliers_false_liberal_use_means/L3-9-cis7-1.txt\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mRead in 18 constraints\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER PackRotamersMover - packrot=======================\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to locate database file /sampling/antibodies/antibody_database_rosetta_design.db\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mReading from: /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrosetta-2019.33+release.1e60c63beb5-py3.6-macosx-10.6-intel.egg/pyrosetta/database//sampling/antibodies/antibody_database_rosetta_design_north_paper.db\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L1-10-1 with 15 datapoints.\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L2-8-1 with 153 datapoints.\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L3-9-cis7-1 with 179 datapoints.\n",
      "\u001b[0mprotocols.antibody.task_operations.AddCDRProfilesOperation: \u001b[0mapplying prob task op\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 5062 rotamers at 61 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating PDInteractionGraph\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m=======================BEGIN MOVER MinMover - minmover=======================\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFGGMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.rosetta_scripts.ParsedProtocol: \u001b[0m\u001b[31m\u001b[1m[ ERROR ]\u001b[0m Exception while processing procotol: \n",
      "\n",
      "File: /Volumes/MacintoshHD3/benchmark/W.fujii.release/rosetta.Fujii.release/_commits_/main/source/src/core/optimization/CartesianMinimizer.cc:66\n",
      "[ ERROR ] UtilityExitException\n",
      "ERROR: Scorefunction not set up for nonideal/Cartesian scoring\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "\n\nFile: /Volumes/MacintoshHD3/benchmark/W.fujii.release/rosetta.Fujii.release/_commits_/main/source/src/core/optimization/CartesianMinimizer.cc:66\n[ ERROR ] UtilityExitException\nERROR: Scorefunction not set up for nonideal/Cartesian scoring\n\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-39-59f74abbd00f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRosettaScriptsParser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mprotocol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenerate_mover_and_apply_to_pose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpose\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"inputs/rabd/ab_design_components.xml\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mprotocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m: \n\nFile: /Volumes/MacintoshHD3/benchmark/W.fujii.release/rosetta.Fujii.release/_commits_/main/source/src/core/optimization/CartesianMinimizer.cc:66\n[ ERROR ] UtilityExitException\nERROR: Scorefunction not set up for nonideal/Cartesian scoring\n\n"
     ]
    }
   ],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    pose = original_pose.clone()\n",
    "    parser = RosettaScriptsParser()\n",
    "    protocol = parser.generate_mover_and_apply_to_pose(pose, \"inputs/rabd/ab_design_components.xml\")\n",
    "    protocol.apply(pose)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Challenge: Custom Design Protocol in code\n",
    "If you want a challenge - try to set these up in-code without RosettaScripts.  It can be tricky - which is why I made PyRosetta finally work optionally with RosettaScripts.  Its good to know how to use both.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-82bd0ff2bc4158ab",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mStarting energy table calculation\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: changing atr/rep split to bottom of energy well\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing lj etables (maxdis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0msmooth_etable: spline smoothing solvation etables (max_dis = 6)\n",
      "\u001b[0mcore.scoring.etable: \u001b[0mFinished calculating energy tables.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/all.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/fd/prepro.ramaProb\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.all.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.gly.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.pro.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/omega/omega_ppdep.valile.txt\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/P_AA_n\n",
      "\u001b[0mcore.scoring.P_AA: \u001b[0mshapovalov_lib::shap_p_aa_pp_smooth_level of 1( aka low_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/P_AA_pp/shapovalov/10deg/kappa131/a20.prop\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mInitializing IdealParametersDatabase with default Ks=300 , 80 , 20 , 10 , 40\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/bondlength_bondangle/default-lengths.txt\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mRead 757 bb-independent lengths.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/bondlength_bondangle/default-angles.txt\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mRead 1456 bb-independent angles.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/bondlength_bondangle/default-torsions.txt\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mRead 1 bb-independent torsions.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/bondlength_bondangle/default-improper.txt\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mRead 2216 bb-independent improper tors.\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.task_operations.ConservativeDesignOperation: \u001b[0mLoading conservative mutational data\n",
      "\u001b[0mcore.import_pose.import_pose: \u001b[0mFile 'malaria_cdrs.pdb' automatically determined to be of type PDB\n",
      "\u001b[0mcore.io.pdb.pdb_reader: \u001b[0mParsing 0 .pdb records with unknown format to search for Rosetta-specific comments.\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m missing heavyatom:  OXT on residue CYS:CtermProteinFull 387\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m missing heavyatom:  OXT on residue ASP:CtermProteinFull 601\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 8 22\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 8 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 22 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 8 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 22 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 24 36\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 24 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 36 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 24 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 36 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 43 58\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 43 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 58 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 43 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 58 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 52 70\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 52 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 70 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 52 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 70 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 72 83\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 72 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 83 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 72 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 83 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 88 98\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 88 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 98 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 88 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 98 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 93 111\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 93 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 111 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 93 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 111 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 113 127\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 113 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 127 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 113 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 127 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 135 146\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 135 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 146 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 135 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 146 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 139 155\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 139 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 155 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 139 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 155 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 157 170\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 157 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 170 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 157 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 170 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 197 261\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 197 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 261 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 197 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 261 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 307 367\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 307 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 367 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 307 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 367 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 407 481\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 407 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 481 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 407 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 481 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 527 582\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 527 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 582 CYS\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 527 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 582 CYD\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SasaMetric - calculating sasa\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SelectedResiduesPyMOLMetric - calculating pymol_selection\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L1_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L2_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L3_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: InteractionEnergyMetric - calculating interaction_energy\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mCreating new peptide-bonded energy container (975)\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/elec_cp_reps.dat\n",
      "\u001b[0mcore.scoring.elec.util: \u001b[0mRead 40 countpair representative atoms\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mshapovalov_lib_fixes_enable option is true.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mshapovalov_lib::shap_dun10_smooth_level of 1( aka lowest_smooth ) got activated.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mBinary rotamer library selected: /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mUsing Dunbrack library binary file '/Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database/rotamer/shapovalov/StpDwn_0-0-0/Dunbrack10.lib.bin'.\n",
      "\u001b[0mcore.pack.dunbrack.RotamerLibrary: \u001b[0mDunbrack 2010 library took 0.252145 seconds to load from binary\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m ################ Cloning pose and Scoring! ##############################\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Ensure that pose is scored\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m before using InteractionEnergyMetric for maximum performance!\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m ##########################################################################\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 11 Omega: TTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_atr 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -35.7698\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -35.7698\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_rep 0.55\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 5415.79\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 2978.68\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_sol 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 34.2565\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 34.2565\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mlk_ball_wtd 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 2.8777\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 2.8777\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_elec 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -8.73098\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -8.73098\n",
      "\u001b[0mcore.scoring.ScoreFunctionFactory: \u001b[0mSCOREFUNCTION: \u001b[32mref2015\u001b[0m\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGrafting CDR: L1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mReturning 16 residues from 195 to 210\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mReverting out-of-date disulfide to thiol type at resid 3\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 891 End: 903 NterO: 3 CterO: 3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mSuperimposing overhang residues\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 11 residues from 892 to 902\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 945\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 945 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 945 CYD\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 10\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 891 -1  EDGE 891 892 1  EDGE 892 964 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 891\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mOriginal End: 903\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mEnd: 902\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mInsert Length: 10\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/centroid_smooth/cen_smooth_params.txt\n",
      "\u001b[0mcore.scoring.ramachandran: \u001b[0mshapovalov_lib::shap_rama_smooth_level of 4( aka highest_smooth ) got activated.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/rama/shapovalov/kappa25/all.ramaProb\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 890  stop: 893  cut: 891  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 900  stop: 903  cut: 901  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.loops.FoldTreeFromLoopsWrapper: \u001b[0mold foldtree FOLD_TREE  EDGE 1 891 -1  EDGE 891 902 1  EDGE 902 974 -1  EDGE 891 892 -2  C    N    EDGE 892 893 -2  C    N    EDGE 893 894 -2  C    N    EDGE 894 895 -2  C    N    EDGE 895 896 -2  C    N    EDGE 896 897 -2  C    N    EDGE 897 898 -2  C    N    EDGE 898 899 -2  C    N    EDGE 899 900 -2  C    N    EDGE 900 901 -2  C    N   \n",
      "New foldtree FOLD_TREE  EDGE 1 889 -1  EDGE 889 891 -1  EDGE 889 894 1  EDGE 894 892 -1  EDGE 894 899 -1  EDGE 899 901 -1  EDGE 899 904 2  EDGE 904 902 -1  EDGE 904 974 -1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  890\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  891\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 892\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 893\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 900\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 901\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 902\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 903\n",
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mFinished initializing centroid residue type set.  Created 62 residue types\n",
      "\u001b[0mcore.chemical.GlobalResidueTypeSet: \u001b[0mTotal time to initialize 0.025629 seconds.\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_distance_score\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_CaCbCb_angle_score\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_CaCbCbCa_dihedral_score\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: scoring/score_functions/disulfides/centroid_backbone_dihedral_score\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mstart 1661.69\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 1\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft Closed early - returning\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m1 1267.29\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mfinish 1267.29\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 151 rotamers at 19 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft meets ideal geometry true\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mSuccess.  Graft closed.\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGrafting CDR: L2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mReturning 14 residues from 219 to 232\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 915 End: 924 NterO: 3 CterO: 3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mSuperimposing overhang residues\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 8 residues from 916 to 923\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 947\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 947 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 947 CYD\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 915\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 916\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 8\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 915 -1  EDGE 915 916 1  EDGE 916 966 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 915\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mOriginal End: 924\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mEnd: 924\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mInsert Length: 8\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 914  stop: 917  cut: 915  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 922  stop: 925  cut: 923  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.loops.FoldTreeFromLoopsWrapper: \u001b[0mold foldtree FOLD_TREE  EDGE 1 915 -1  EDGE 915 924 1  EDGE 924 974 -1  EDGE 915 916 -2  C    N    EDGE 916 917 -2  C    N    EDGE 917 918 -2  C    N    EDGE 918 919 -2  C    N    EDGE 919 920 -2  C    N    EDGE 920 921 -2  C    N    EDGE 921 922 -2  C    N    EDGE 922 923 -2  C    N   \n",
      "New foldtree FOLD_TREE  EDGE 1 913 -1  EDGE 913 915 -1  EDGE 913 918 1  EDGE 918 916 -1  EDGE 918 921 -1  EDGE 921 923 -1  EDGE 921 926 2  EDGE 926 924 -1  EDGE 926 974 -1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 915\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  914\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  915\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 916\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 916\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 917\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 922\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 923\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 924\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 925\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mstart 1669.58\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 1\n",
      "\u001b[0mprotocols.loops.util: \u001b[0mPepBondGeom: 923 - 72 L  Ca-C-N --  99.5 (min): 142.34 : 134.5 (max)\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m1 1306.25\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 2\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft Closed early - returning\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m2 1219.92\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mfinish 1219.92\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 149 rotamers at 14 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft meets ideal geometry true\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mSuccess.  Graft closed.\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mGrafting CDR: L3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mReturning 15 residues from 259 to 273\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mReverting out-of-date disulfide to thiol type at resid 3\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 955 End: 965 NterO: 3 CterO: 3\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mSuperimposing overhang residues\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mDeleting 9 residues from 956 to 964\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 771 845\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 771 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 845 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mFound disulfide between residues 891 955\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 955 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 891 CYD\n",
      "\u001b[0mcore.conformation.Conformation: \u001b[0mcurrent variant for 955 CYD\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_point 955\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_start 956\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_end 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0minsert_length 9\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mFOLD_TREE  EDGE 1 955 -1  EDGE 955 956 1  EDGE 956 965 -1\n",
      "\u001b[0mprotocols.grafting.GraftMoverBase: \u001b[0mInsertion complete.\n",
      "\u001b[0mprotocols.grafting.AnchoredGraftMover: \u001b[0mSetting default movemap\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mStart: 955\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mOriginal End: 965\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mEnd: 965\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mInsert Length: 9\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 954  stop: 957  cut: 955  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mLOOP start: 963  stop: 966  cut: 964  size: 4  skip rate: 0  extended?: False\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m\n",
      "\u001b[0mprotocols.loops.FoldTreeFromLoopsWrapper: \u001b[0mold foldtree FOLD_TREE  EDGE 1 955 -1  EDGE 955 965 1  EDGE 965 974 -1  EDGE 955 956 -2  C    N    EDGE 956 957 -2  C    N    EDGE 957 958 -2  C    N    EDGE 958 959 -2  C    N    EDGE 959 960 -2  C    N    EDGE 960 961 -2  C    N    EDGE 961 962 -2  C    N    EDGE 962 963 -2  C    N    EDGE 963 964 -2  C    N   \n",
      "New foldtree FOLD_TREE  EDGE 1 953 -1  EDGE 953 955 -1  EDGE 953 958 1  EDGE 958 956 -1  EDGE 958 962 -1  EDGE 962 964 -1  EDGE 962 967 2  EDGE 967 965 -1  EDGE 967 974 -1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 1\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 955\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mLoop: 2\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mAdd variant to: 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  954\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized  955\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 956\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 956\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 957\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 963\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0mideal 964\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 965\n",
      "\u001b[0mprotocols.grafting.util: \u001b[0midealized 966\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mstart 1612.44\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mround 1\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft Closed early - returning\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0m1 1219.16\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mfinish 1219.16\n",
      "\u001b[0mcore.pack.task: \u001b[0mPacker task: initialize from command line()\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 118 rotamers at 17 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mGraft meets ideal geometry true\n",
      "\u001b[0mprotocols.grafting.CCDEndsGraftMover: \u001b[0mComplete\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mChecking Geometry\n",
      "\u001b[0mprotocols.antibody.AntibodyCDRGrafter: \u001b[0mSuccess.  Graft closed.\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.constraints.CDRDihedralConstraintMover: \u001b[0mL1-10-1 data: 42 cutoff: 10\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mread constraints from /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database/sampling/antibodies/cluster_based_constraints/CircularHarmonic/outliers_false_liberal_use_means/L1-10-1.txt\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mRead in 20 constraints\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.constraints.CDRDihedralConstraintMover: \u001b[0mL2-8-1 data: 632 cutoff: 10\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mread constraints from /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database/sampling/antibodies/cluster_based_constraints/CircularHarmonic/outliers_false_liberal_use_means/L2-8-1.txt\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mRead in 16 constraints\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.antibody.constraints.CDRDihedralConstraintMover: \u001b[0mL3-9-cis7-1 data: 419 cutoff: 10\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mread constraints from /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database/sampling/antibodies/cluster_based_constraints/CircularHarmonic/outliers_false_liberal_use_means/L3-9-cis7-1.txt\n",
      "\u001b[0mcore.scoring.constraints.ConstraintsIO: \u001b[0mRead in 18 constraints\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to locate database file /sampling/antibodies/antibody_database_rosetta_design.db\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mReading from: /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database//sampling/antibodies/antibody_database_rosetta_design_north_paper.db\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L1-10-1 with 15 datapoints.\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L2-8-1 with 153 datapoints.\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L3-9-cis7-1 with 179 datapoints.\n",
      "\u001b[0mprotocols.antibody.task_operations.AddCDRProfilesOperation: \u001b[0mapplying prob task op\n",
      "\u001b[0mcore.scoring.CartesianBondedEnergy: \u001b[0mCreating new peptide-bonded energy container (974)\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 0 rotamers at 0 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 771 BRANCH 1\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 845 BRANCH 1\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 891 BRANCH 1\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 955 BRANCH 1\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to locate database file /sampling/antibodies/antibody_database_rosetta_design.db\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mReading from: /Users/jadolfbr/Library/Python/3.6/lib/python/site-packages/pyrosetta-2019.39+release.93456a567a8-py3.6-macosx-10.6-intel.egg/pyrosetta/database//sampling/antibodies/antibody_database_rosetta_design_north_paper.db\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L1-10-1 with 15 datapoints.\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L2-8-1 with 153 datapoints.\n",
      "\u001b[0mantibody.database.AntibodyDatabaseManager: \u001b[0mLoaded L3-9-cis7-1 with 179 datapoints.\n",
      "\u001b[0mprotocols.antibody.task_operations.AddCDRProfilesOperation: \u001b[0mapplying prob task op\n",
      "\u001b[0mcore.pack.pack_rotamers: \u001b[0mbuilt 0 rotamers at 0 positions.\n",
      "\u001b[0mcore.pack.interaction_graph.interaction_graph_factory: \u001b[0mInstantiating DensePDInteractionGraph\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 771 BRANCH 1\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 845 BRANCH 1\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 891 BRANCH 1\n",
      "\u001b[0mcore.pose.util: \u001b[0m\u001b[1m[ WARNING ]\u001b[0m Unable to find atom_tree atom for this Rosetta branch connection angle: residue 955 BRANCH 1\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SasaMetric - calculating sasa\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SelectedResiduesPyMOLMetric - calculating pymol_selection\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L1_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L2_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: SequenceMetric - calculating L3_sequence\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mprotocols.analysis.simple_metrics.RunSimpleMetricsMover: \u001b[0mRunning: InteractionEnergyMetric - calculating interaction_energy\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mbasic.io.database: \u001b[0mDatabase file opened: sampling/antibodies/cluster_center_dihedrals.txt\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody numbering scheme definitions read successfully\n",
      "\u001b[0mprotocols.antibody.AntibodyNumberingParser: \u001b[0mAntibody CDR definition read successfully\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSuccessfully finished the CDR definition\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Detecting Regular CDR H3 Stem Type\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSRWGGDGFYAMDYW\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: KINKED\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mAC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for H3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 13 Omega: TTTTTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L1\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 10 Omega: TTTTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L2\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 8 Omega: TTTTTTTT\n",
      "\u001b[0mantibody.AntibodyInfo: \u001b[0mSetting up CDR Cluster for L3\n",
      "\u001b[0mprotocols.antibody.cluster.CDRClusterMatcher: \u001b[0mLength: 9 Omega: TTTTTTCTT\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_atr 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -21.8976\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -21.8976\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_rep 0.55\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 227.209\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 124.965\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_sol 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 20.1315\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 20.1315\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mlk_ball_wtd 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: 1.54267\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: 1.54267\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0m\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mfa_elec 1\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mUnweighted: -0.883278\n",
      "\u001b[0mcore.simple_metrics.metrics.InteractionEnergyMetric: \u001b[0mWeighted: -0.883278\n"
     ]
    }
   ],
   "source": [
    "### BEGIN SOLUTION\n",
    "from rosetta.protocols.antibody.residue_selector import *\n",
    "from rosetta.protocols.antibody.task_operations import *\n",
    "from rosetta.protocols.antibody.constraints import *\n",
    "from rosetta.protocols.antibody import *\n",
    "from rosetta.core.simple_metrics.metrics import *\n",
    "from rosetta.core.simple_metrics.per_residue_metrics import *\n",
    "from rosetta.core.select.movemap import *\n",
    "from rosetta.protocols.analysis.simple_metrics import RunSimpleMetricsMover\n",
    "from rosetta.core.pack.task import *\n",
    "\n",
    "\n",
    "####################\n",
    "## Clone our pose ##\n",
    "####################\n",
    "\n",
    "pose = original_pose.clone()\n",
    "ab_info = AntibodyInfo(pose)\n",
    "\n",
    "#########################################################\n",
    "## Setup Scorefunction since we are not using cmd-line ##\n",
    "#########################################################\n",
    "cart = create_score_function(\"ref2015_cart\")\n",
    "\n",
    "\n",
    "#############################\n",
    "## Setup Residue Selectors ##\n",
    "#############################\n",
    "#1\t\t<CDR name=\"light_cdrs\" cdrs=\"L1,L2,L3\" />\n",
    "#2 \t\t<CDR name=\"L1\" cdrs=\"L1\"/>\n",
    "#3 \t\t<CDR name=\"L2\" cdrs=\"L2\"/>\n",
    "#4 \t\t<CDR name=\"L3\" cdrs=\"L3\"/>\n",
    "#5 \t\t<AntibodyRegion name=\"antigen\" region=\"antigen_region\" />\n",
    "#6 \t\t<AntibodyRegion name=\"framework\" region=\"framework_region\" />\n",
    "#7 \t\t<AntibodyRegion name=\"cdrs\" region=\"cdr_region\" />\n",
    "\n",
    "#1 <CDR name=\"light_cdrs\"\n",
    "light_sele = CDRResidueSelector()\n",
    "light_list = vector1_protocols_antibody_CDRNameEnum()\n",
    "[light_list.append(x) for x in [l1, l2, l3]]\n",
    "\n",
    "light_sele.set_cdrs(light_list)\n",
    "\n",
    "#2 <CDR name=\"L1\"\n",
    "l1_sele = CDRResidueSelector()\n",
    "l1_sele.set_cdr(l1)\n",
    "\n",
    "#3 <CDR name=\"L2\"\n",
    "l2_sele = CDRResidueSelector()\n",
    "l2_sele.set_cdr(l2)\n",
    "\n",
    "#4 <CDR name=\"L3\"\n",
    "l3_sele = CDRResidueSelector()\n",
    "l3_sele.set_cdr(l3)\n",
    "\n",
    "#5 <AntibodyRegion name=\"antigen\"\n",
    "antigen_sele = AntibodyRegionSelector()\n",
    "antigen_sele.set_region(antigen_region)\n",
    "\n",
    "#6 <AntibodyRegion name=\"framework\"\n",
    "framework_sele = AntibodyRegionSelector()\n",
    "framework_sele.set_region(framework_region)\n",
    "\n",
    "#7 <AntibodyRegion name=\"cdrs\"\n",
    "cdrs_sele = AntibodyRegionSelector()\n",
    "cdrs_sele.set_region(cdr_region)\n",
    "\n",
    "\n",
    "##########################\n",
    "## Setup Simple Metrics ##\n",
    "##########################\n",
    "#1\t\t<SasaMetric name=\"sasa\" residue_selector=\"light_cdrs\" />\n",
    "#2\t\t<SelectedResiduesPyMOLMetric name=\"cdr_selection\" residue_selector=\"light_cdrs\" />\n",
    "#3\t\t<SequenceMetric name=\"L1_seq\" residue_selector=\"L1\" custom_type=\"L1\"/>\n",
    "#4\t\t<SequenceMetric name=\"L2_seq\" residue_selector=\"L2\" custom_type=\"L2\"/>\n",
    "#4\t\t<SequenceMetric name=\"L3_seq\" residue_selector=\"L3\" custom_type=\"L3\"/>\n",
    "#5\t\t<InteractionEnergyMetric name=\"cdr-int\" residue_selector=\"light_cdrs\" residue_selector2=\"antigen\" />\n",
    "\n",
    "#1 <SasaMetric name=\"sasa\"\n",
    "sasa = SasaMetric(light_sele)\n",
    "\n",
    "#2 <SelectedResiduesPyMOLMetric name=\"cdr_selection\"\n",
    "cdr_selection = SelectedResiduesPyMOLMetric(light_sele)\n",
    "L1_seq = SequenceMetric(l1_sele)\n",
    "L1_seq.set_custom_type(\"L1\")\n",
    "\n",
    "#3 <SequenceMetric name=\"L1_seq\"\n",
    "L2_seq = SequenceMetric(l2_sele)\n",
    "L2_seq.set_custom_type(\"L2\")\n",
    "\n",
    "#4 <SequenceMetric name=\"L2_seq\"\n",
    "L3_seq = SequenceMetric(l3_sele)\n",
    "L3_seq.set_custom_type(\"L3\")\n",
    "\n",
    "#5 <InteractionEnergyMetric name=\"cdr-int\"\n",
    "cdr_int = InteractionEnergyMetric(light_sele, antigen_sele)\n",
    "cdr_int.set_scorefunction(cart)\n",
    "\n",
    "\n",
    "########################## \n",
    "## Setup TaskOperations ##\n",
    "##########################\n",
    "#1\t \t<RestrictToCDRsAndNeighbors name=\"restrict_to_cdrs\" cdrs=\"L1,L2,L3\" design_framework=\"1\" design_cdrs=\"1\"/>\n",
    "#2\t \t<AddCDRProfilesOperation name=\"profiles\" cdrs=\"L1,L2,L3\" fallback_strategy=\"CONSERVATIVE\" /> \n",
    "\n",
    "light_vec_bool = vector1_bool(6)\n",
    "light_vec_bool[l1] = True\n",
    "light_vec_bool[l2] = True\n",
    "light_vec_bool[l3] = True\n",
    "\n",
    "#1 <RestrictToCDRsAndNeighbors name=\"restrict_to_cdrs\"\n",
    "restrict_to_cdrs = RestrictToCDRsAndNeighbors()\n",
    "restrict_to_cdrs.set_cdrs(light_vec_bool)\n",
    "restrict_to_cdrs.set_allow_design_neighbor_framework(True)\n",
    "restrict_to_cdrs.set_allow_design_cdr(True)\n",
    "\n",
    "#2 <AddCDRProfilesOperation name=\"profiles\"\n",
    "profiles = AddCDRProfilesOperation()\n",
    "profiles.set_cdrs(light_vec_bool)\n",
    "profiles.set_fallback_strategy(design.seq_design_conservative)\n",
    "\n",
    "\n",
    "##########################\n",
    "## Setup MovemapFactory ##\n",
    "##########################\n",
    "#1\t\t<MoveMapFactory name=\"movemap_cdrs\" bb=\"0\" chi=\"0\">\n",
    "#1\t\t\t<Backbone residue_selector=\"light_cdrs\" />\n",
    "#1\t\t\t<Chi residue_selector=\"light_cdrs\" />\n",
    "#1\t\t</MoveMapFactory>\n",
    "\n",
    "#1\n",
    "mmf = MoveMapFactory()\n",
    "mmf.all_bb(False) #First, disable everything, then enable from actions.\n",
    "mmf.all_chi(False)\n",
    "mmf.add_bb_action(mm_enable, light_sele)\n",
    "mmf.add_chi_action(mm_enable, light_sele)\n",
    "mmf.set_cartesian(True)\n",
    "\n",
    "\n",
    "##################\n",
    "## Setup Movers ##\n",
    "##################\n",
    "#1\t\t<CDRDihedralConstraintMover name=\"dih_mover_L1\" cdr=\"L1\" use_cluster_csts=\"1\" />\n",
    "#2\t\t<CDRDihedralConstraintMover name=\"dih_mover_L2\" cdr=\"L2\" use_cluster_csts=\"1\" />\n",
    "#3\t\t<CDRDihedralConstraintMover name=\"dih_mover_L3\" cdr=\"L3\" use_cluster_csts=\"1\" />\n",
    "#4\t\t<AntibodyCDRGrafter name=\"grafter\" cdrs=\"L1,L2,L3\" input_ab_scheme=\"AHO_Scheme\" cdr_definition=\"North\" donor_structure_from_pdb=\"malaria_cdrs.pdb\" use_secondary_graft_mover=\"1\" optimize_cdrs=\"0\" optimize_cdr4_if_neighbor=\"1\" />\n",
    "#5\t\t<PackRotamersMover name=\"packrot\" task_operations=\"restrict_to_cdrs,profiles\" />\n",
    "\n",
    "#\t\tMinimize Light chain CDRs using the MoveMapFactory.  We do it in cartesian to prevent lever-arm effects\n",
    "#6\t\t<MinMover name=\"minmover\" cartesian=\"1\" movemap_factory=\"movemap_cdrs\" tolerance=\".1\"/>\n",
    "\n",
    "#\t\tRun our metrics\n",
    "#7\t\t<RunSimpleMetrics name=\"design_metrics\" metrics=\"sasa,cdr_selection,L1_seq,L2_seq,L3_seq,cdr-int\" prefix=\"design_\" />\n",
    "#8\t\t<RunSimpleMetrics name=\"native_metrics\" metrics=\"sasa,cdr_selection,L1_seq,L2_seq,L3_seq,cdr-int\"  prefix=\"native_\" />\n",
    "\n",
    "\n",
    "#1 <CDRDihedralConstraintMover name=\"dih_mover_L1\"\n",
    "dih_csts_l1 = CDRDihedralConstraintMover()\n",
    "dih_csts_l1.set_cdr(l1)\n",
    "\n",
    "#2 <CDRDihedralConstraintMover name=\"dih_mover_L2\"\n",
    "dih_csts_l2 = CDRDihedralConstraintMover()\n",
    "dih_csts_l2.set_cdr(l2)\n",
    "\n",
    "#3 <CDRDihedralConstraintMover name=\"dih_mover_L3\"\n",
    "dih_csts_l3 = CDRDihedralConstraintMover()\n",
    "dih_csts_l3.set_cdr(l3)\n",
    "\n",
    "#4 <AntibodyCDRGrafter name=\"grafter\"\n",
    "new_pose = pose_from_pdb(\"malaria_cdrs.pdb\")\n",
    "\n",
    "grafter = AntibodyCDRGrafter(ab_info)\n",
    "grafter.set_donor_structure(new_pose)\n",
    "grafter.set_cdrs(light_vec_bool)\n",
    "grafter.set_use_secondary_graft_mover_if_needed(True)\n",
    "grafter.set_optimize_cdrs(False)\n",
    "grafter.set_include_cdr4_in_optimization(True)\n",
    "\n",
    "#5 <PackRotamersMover name=\"packrot\"\n",
    "tf = TaskFactory()\n",
    "tf.push_back(restrict_to_cdrs)\n",
    "tf.push_back(profiles)\n",
    "\n",
    "packer = PackRotamersMover()\n",
    "packer.task_factory(tf)\n",
    "packer.score_function(cart)\n",
    "\n",
    "#6 <MinMover name=\"minmover\"\n",
    "minmover = MinMover()\n",
    "minmover.cartesian(True)\n",
    "minmover.movemap_factory(mmf)\n",
    "minmover.tolerance(.1)\n",
    "minmover.score_function(cart)\n",
    "\n",
    "#7 <RunSimpleMetrics name=\"design_metrics\" metrics=\"sasa,cdr_selection,L1_seq,L2_seq,L3_seq,cdr-int\"\n",
    "metrics = [sasa, cdr_selection, L1_seq, L2_seq, L3_seq, cdr_int]\n",
    "\n",
    "design_metrics = RunSimpleMetricsMover()\n",
    "design_metrics.set_prefix(\"design_\")\n",
    "[design_metrics.add_simple_metric(metric) for metric in metrics] #Save a few lines as the vector1 for metrics is not yet bound to PyRosetta.\n",
    "\n",
    "#8 <RunSimpleMetrics name=\"native_metrics\" metrics=\"sasa,cdr_selection,L1_seq,L2_seq,L3_seq,cdr-int\"\n",
    "native_metrics = RunSimpleMetricsMover()\n",
    "native_metrics.set_prefix(\"native_\")\n",
    "[native_metrics.add_simple_metric(metric) for metric in metrics]\n",
    "\n",
    "\n",
    "######################\n",
    "## Run The Protocol ##\n",
    "######################\n",
    "#You also use a MoverContainer from previous tutorials here if you wanted to.\n",
    "\n",
    "#1\t\tRun Metrics on the native antibody\n",
    "#\t\t<Add mover_name=\"native_metrics\" />\n",
    "\n",
    "#2\t\tGraft Light chain CDRs from our donor into our current antibody.\n",
    "#\t\t<Add mover_name=\"grafter\" />\n",
    "\n",
    "#3\t\tAdd Constraints to the CDRs so that min doesn't change them too much\n",
    "#\t\t<Add mover_name=\"dih_mover_L1\" />\n",
    "#\t\t<Add mover_name=\"dih_mover_L2\" />\n",
    "#\t\t<Add mover_name=\"dih_mover_L3\" />\n",
    "\n",
    "#4\t\tDesign framework around light chain to accomodate the grafted CDRs.  Pack/Min/Pack\n",
    "#\t\t<Add mover_name=\"packrot\" />\n",
    "#\t\t<Add mover_name=\"minmover\" />\n",
    "#\t\t<Add mover_name=\"packrot\" /> \n",
    "#\t\t<Add mover_name=\"minmover\" />\n",
    "\n",
    "#5\t\tRun Simple Metrics on the result\n",
    "#\t\t<Add mover_name=\"design_metrics\" />\n",
    "\n",
    "\n",
    "if not os.getenv(\"DEBUG\"):\n",
    "    #1\t\tRun Metrics on the native antibody\n",
    "    native_metrics.apply(pose)\n",
    "\n",
    "    #2\t\tGraft Light chain CDRs from our donor into our current antibody.\n",
    "    grafter.apply(pose)\n",
    "\n",
    "    #3\t\tAdd Constraints to the CDRs so that min doesn't change them too much\n",
    "    dih_csts_l1.apply(pose)\n",
    "    dih_csts_l2.apply(pose)\n",
    "    dih_csts_l3.apply(pose)\n",
    "\n",
    "    #4\t\tDesign framework around light chain to accomodate the grafted CDRs.  Pack/Min/Pack\n",
    "    packer.apply(pose)\n",
    "    minmover.apply(pose)\n",
    "    packer.apply(pose)\n",
    "    minmover.apply(pose)\n",
    "\n",
    "    #5\t\tRun Simple Metrics on the result\n",
    "    design_metrics.apply(pose)\n",
    "\n",
    "### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tutorial B:  Optimizing Interface Energy (opt-dG)\n",
    "\n",
    "### Tut B1. Optimizing dG\n",
    "\n",
    "Here, we want to set the protocol to optimize the interface energy during Monte Carlo instead of total energy.  The interface energy is calculated by the InterfaceAnalyzerMover through a specialized MonteCarlo object called `MonteCarloInterface`.  This is useful to improve binding energy and generally results in better interface energies .  Resulting models should still be pruned for high total energy.  This was benchmarked in the paper, and has been used for real-life designs to the HIV epitope (165 seconds for 1 decoy).\n",
    "\n",
    "Use the provided XML or set this up through code. \n",
    "\n",
    "    <AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" graft_design_cdrs=\"L1,L3\" mc_optimize_dG=\"1\" />\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-7664283d2dba2e99",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    ### BEGIN SOLUTION\n",
    "\n",
    "    pose = original_pose.clone()\n",
    "    rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" graft_design_cdrs=\"L1,L3\" seq_design_cdrs=\"L1,L3\" mc_optimize_dG=\"1\" light_chain=\"kappa\"/>')\n",
    "    rabd.apply(pose)\n",
    "\n",
    "    ### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the scorefile with nstruct=5 from `expected_outputs/rabd/tutB1_score.sc`\n",
    "\n",
    "Compare this data to tutorial A2.  Are the interface energies better?  Has the Shape Complementarity improved (sc score) improved? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut B2. Optimizing Interface Energy and Total Score (opt-dG and opt-E)\n",
    "\n",
    "Here, we want to set the protocol to optimize the interface energy during Monte Carlo, but we want to add some total energy to the weight.  Because the overall numbers of total energy will dominate the overall numbers, we only add a small weight for total energy.  This has not been fully benchmarked, but if your models have very bad total energy when using opt-dG - consider using it.  (178 sec for 1 nstruct)\n",
    "\n",
    "\n",
    "    <AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" graft_design_cdrs=\"L1,L3\" mc_optimize_dG=\"1\" mc_total_weight=\".01 mc_interface_weight=\".99 light_chain=\"kappa\"/>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-119248d958256181",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    ### BEGIN SOLUTION\n",
    "    pose = original_pose.clone()\n",
    "    rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3\" graft_design_cdrs=\"L1,L3\" mc_optimize_dG=\"1\" mc_total_weight=\".01 mc_interface_weight=\".99 light_chain=\"kappa\"/>')\n",
    "    rabd.apply(pose)\n",
    "    ### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the scorefile from an nstruct=5 run to compare total energies (`total_score`) of this run vs the one right before it located at `expected_outputs/rabd/tutB2_score.sc`.  Are the total scores better?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tutorial C: Towards DeNovo Design: Integrated Dock/Design\n",
    "\n",
    "This tutorial takes a long time to run as docking is fairly slow - even with the optimizations that are part of RAbD.  PLEASE USE THE PREGENERATED OUTPUT.  The top 10 designs from each tutorial and associated scorefiles of a 1000 nstruct cluster run are in the output directory.  Note that we are starting these tutorials with a pre-relaxed structure in order to get more reliable rosetta energies.  Since we are running a large nstruct, we will escape the local energy well that this leads us into.\n",
    "\n",
    "### Tut C1. RosettaDocking\n",
    "\n",
    "In this example, we use integrated RosettaDock (with sequence design during the high-res step) to sample the antibody-antigen orientation, but we don't care where the antibody binds to the antigen.  Just that it binds. IE - No Constraints. The RAbD protocol always has at least Paratope SiteConstraints enabled to make sure any docking is contained to the paratope (like most good docking programs).\n",
    "\n",
    "This takes too long to run, so PLEASE USE THE OUTPUT GENERATED FOR YOU. We will use opt-dG here and for these tutorials, we will be sequence-designing all cdrs to begin to create a better interface.  Note that sequence design happens whereever packing occurs - Including during high-resolution docking.\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" mc_optimize_dG=\"1\" do_dock=\"1\" seq_design_cdrs=\"L1,L2,L3,H1,H2,H3\" graft_design_cdrs=\"L1,L2,L3,H1,H2\" light_chain=\"kappa\"/>\n",
    "\n",
    "**Use pymol to load the files, and load tutC1_score from the expected_outputs directory as a pandas dataframe.**\n",
    "\n",
    "\t\tpymol my_ab.pdb expected_outputs/rabd/top10_C1/* \n",
    "\t\t@color_cdrs.pml\n",
    "\t\tcenter full_epitope\n",
    "\n",
    "Where is the antibody in the resulting designs?  Are the interfaces restricted to the Paratope?  Has the epitope moved relative to the starting interface?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut C2. Auto Epitope Constraints\n",
    "\n",
    "Allow Dock-Design, incorporating auto-generated SiteConstraints to keep the antibody around the starting interface residues.  These residues are determined by being within 6A to the CDR residues (This interface distance can be customized).  Again, these results are provided for you.\n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" mc_optimize_dG=\"1\" do_dock=\"1\" use_epitope_csts=\"1\" \n",
    "\t   seq_design_cdrs=\"L1,L2,L3,H1,H2,H3\" graft_design_cdrs=\"L1,L2,L3,H1,H2\" light_chain=\"kappa\"/>\n",
    "\n",
    "**Use pymol to load the files and checkout the scores in `expected_outputs/rabd/tutC2_score.sc` as before.**\n",
    "\n",
    "\n",
    "\tpymol my_ab.pdb expected_outputs/rabd/top10_C2/* \n",
    "\t@color_cdrs.pml\n",
    "\tcenter full_epitope\n",
    "\n",
    "How do these compare with with the previous tutorial?  Are the antibodies closer to the starting interface?  Are the scores better?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut C3. Specific Residue Epitope Constraints\n",
    "\n",
    "Allow Dock-Design, as above, but specify the Epitope Residues and Paratope CDRs to guide the dock/design to have these interact.\n",
    "\n",
    "For now, we are more focused on the light chain. We could do this as a two-stage process, where we first optimize positioning and CDRs of the light chain and then the heavy chain or simply add heavy chain CDRs to the paratope CDRs option. \n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" mc_optimize_dG=\"1\" do_dock=\"1\" use_epitope_csts=\"1\" \n",
    "\t\tepitope_residues=\"38J,52J,34K,37K\" paratope_cdrs=\"L1,L3\" \n",
    "\t\tseq_design_cdrs=\"L1,L2,L3,H1,H2,H3\" graft_design_cdrs=\"L1,L2,L3,H1,H2\" light_chain=\"kappa\"/>\n",
    "\n",
    "**Again, load these into Pymol and take a look at the scorefile in a dataframe.**  \n",
    "\n",
    "\tpymol my_ab.pdb expected_outputs/rabd/top10_C3/* \n",
    "\t@color_cdrs.pml\n",
    "\tcenter full_epitope\n",
    "\n",
    "Now that we have specified where we want the interface to be and are additionally designing more CDRS, how do the enegies compare?  Are we starting to get a decent interface with the lowest energy structure?\n",
    "\n",
    "How do these compare with the previous runs?  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tutorial D: Advanced Settings and CDR Instruction File Customization\n",
    "\n",
    "Once again, all output files are in `expected_outputs`.  Please use these if you want - as many of these take around 10 minutes to run.\n",
    "\n",
    "### Tut D1.  CDR Instruction File\n",
    "\n",
    "More complicated design runs can be created by using the Antibody Design Instruction file.  This file allows complete customization of the design run. See below for a review of the syntax of the file and possible customization.  An instruction file is provided where we use conservative design on L1 and graft in L1, H2, and H1 CDRs at a longer length to attempt to create a larger interface area.  More info on instruction file syntax can be found at the end of this tutorial. (150 seconds on my laptop for nstruc 1)\n",
    "\n",
    "\tcp ../inputs/my_instruction_file.txt .\n",
    "\tcp ../inputs/default_cdr_instructions.txt .\n",
    "\n",
    "\n",
    "Take a look at the default CDR instructions.  These are loaded by default into Rosetta.  There is syntax examples at the end of the file. Run the XML or attempt to use it in-code.\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" instruction_file=\"my_instruction_file.txt\" \n",
    "\t   seq_design_cdrs=\"L1,L3,H1,H2,H3\" graft_design_cdrs=\"L1,H2,H1\" random_start=\"1\" light_chain=\"kappa\"/>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-41c91bb1703d613c",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    ### BEGIN SOLUTION\n",
    "\n",
    "    pose = original_pose.clone()\n",
    "    rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" instruction_file=\"my_instruction_file.txt\" seq_design_cdrs=\"L1,L3,H1,H2,H3\" graft_design_cdrs=\"L1,H2,H1\" random_start=\"1\" light_chain=\"kappa\"/>')\n",
    "    rabd.apply(pose)\n",
    "\n",
    "    ### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut D2.  Dissallow AAs and the Resfile\n",
    "\n",
    "Here, we will disallow ANY sequence design into Proline residues and Cysteine residues, while giving a resfile to further LIMIT design and packing as specific positions. These can be given as 3 or 1 letter codes and mixed codes such as PRO and C are accepted. Note that the resfile does NOT turn any residues ON, it is simply used to optionally LIMIT design residue types and design and packing positions. \n",
    "    \n",
    "Resfile syntax can be found here: [https://www.rosettacommons.org/docs/wiki/rosetta_basics/file_types/resfiles] Note that specifying a resfile and dissalowing aa are only currently available as cmd-line options that are read by RAbD.\n",
    "\n",
    "Runtime is less than a minute for nstruct 1.\n",
    "\n",
    "\tcp ../inputs/rabd/my_resfile.resfile .\n",
    "\n",
    "Take a look at the resfile.  Can you describe what it is we are doing with it?\n",
    "Unfortunately, at the moment, resfile setting is only available as a cmd-line option that needs to be set in the `init()` function as `-resfile my_resfile.resfile`\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3,H1,H2,H3\" light_chain=\"kappa\"/>\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut D3. Mintype\n",
    "\n",
    "Here, we will change the mintype to relax.  This mintype enables Flexible-Backbone design as we have seen in previous workshops.  Our default is to use min/pack cycles, but relax typically works better.  However, it also takes considerably more time!  This tutorial takes about 339 seconds for one struct!\n",
    "\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3,H1,H3\" mintype=\"relax light_chain=\"kappa\"/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "cell-a228ebfe14b610a7",
     "locked": false,
     "points": 0,
     "schema_version": 3,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "if not os.getenv(\"DEBUG\"):\n",
    "    ### BEGIN SOLUTION\n",
    "\n",
    "    pose = original_pose.clone()\n",
    "    rabd = XmlObjects.static_get_mover('<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3,H1,H3\" mintype=\"relax\" light_chain=\"kappa\"/>')\n",
    "    rabd.apply(pose)\n",
    "\n",
    "    ### END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut D4. Framework Design\n",
    "\n",
    "Finally, we want to allow the framework residues AROUND the CDRs we will be designing and any interacting antigen residues to design as well here. In addition, we will disable conservative framework design as we want something funky (this is not typically recommended and is used here to indicate what you CAN do.  Note that we will also design the interface of the antigen using the `-design_antigen` option.  This can be useful for vaccine design. Note that these design options are cmd-line ony options currently (but will be available in a later version of Rosetta). Approx 900 second runtime.\n",
    "\n",
    "\n",
    "\tantibody_designer.linuxgccrelease -s my_ab.pdb -seq_design_cdrs L1 L3 H1 H3 \\\n",
    "\t\t\t    -light_chain kappa -resfile my_resfile.resfile -disallow_aa PRO CYS \\\n",
    "\t\t\t    -mintype relax -design_antigen -design_framework \\\n",
    "\t\t\t    -conservative_framework_design false -nstruct 1 -out:prefix tutD4_\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3,H1,H3\" mintype=\"relax\" light_chain=\"kappa\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tut D5.  H3 Stem, kT, and Sequence variablility.\n",
    "\n",
    "Finally, we want increased variability for our sequence designs.  So, we will increase number of sampling rounds for our lovely cluster profiles using the `-seq_design_profile_samples` option, increase kT, and allow H3 stem design.  \n",
    "\n",
    "We will enable H3 Stem design here, which can cause a flipping of the H3 stem type from bulged to non-bulged and vice-versa.  Typically, if you do this, you may want to run loop modeling on the top designs to confirm the H3 structure remains in-tact. Note that once again, these sequence-design specific options must be set on the  cmd-line.  \n",
    "\t\n",
    "Description of the `seq_design_profile_samples` option (default 1): \"If designing using profiles, this is the number of times the profile is sampled each time packing done.  Increase this number to increase variability of designs - especially if not using relax as the mintype.\"\n",
    "\n",
    "This tutorial takes approx 450 seconds.\n",
    "\n",
    "\tantibody_designer.linuxgccrelease -s my_ab.pdb -seq_design_cdrs L1 L2 H3 \\\n",
    "\t\t\t    -graft_design_cdrs L1 L2 -light_chain kappa -design_H3_stem -inner_kt 2.0 \\\n",
    "\t\t\t    -outer_kt 2.0 -seq_design_profile_samples 5 -nstruct 5 -out:prefix tutD5_\n",
    "\n",
    "\t<AntibodyDesignMover name=\"RAbD\" seq_design_cdrs=\"L1,L3,H1,H3\" mintype=\"relax\" \n",
    "\t\t\t                                                        inner_kt=\"2.0\" outer_kt=\"2.0\"/>\n",
    "\n",
    "How different is the sequence of L1,L2, and H3 from our starting antibody?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__You should now be ready to explore and use RosettaAntibodyDesign on your own.  Congrats!  Thanks for going through this tutorial!__\n",
    "\n",
    "The full reference manual can be found here: https://www.rosettacommons.org/docs/latest/application_documentation/antibody/RosettaAntibodyDesign#antibody-design-cdr-instruction-file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [Side Chain Conformations and Dunbrack Energies](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.01-Side-Chain-Conformations-and-Dunbrack-Energies.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [Protein Design with a Resfile and FastRelax](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.03-Design-with-a-resfile-and-relax.ipynb) ><p><a href=\"https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.02-Packing-design-and-regional-relax.ipynb\"><img align=\"left\" src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open in Colab\" title=\"Open in Google Colaboratory\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--NAVIGATION-->\n",
    "< [RosettaAntibody Framework](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.01-RosettaAntibody-Framework-and-SimpleMetrics.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [RosettaCarbohydrates](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/13.00-RosettaCarbohydrates-Working-with-Glycans.ipynb) ><p><a href=\"https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.02-RosettaAntibodyDesign-RAbD.ipynb\"><img align=\"left\" src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open in Colab\" title=\"Open in Google Colaboratory\"></a>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Create Assignment",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
